// angular4-weui/es2015/index.es5.js
import * as tslib_1 from "tslib";
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
import { ApplicationRef, Component, ComponentFactoryResolver, ContentChild, ContentChildren, Directive, ElementRef, EventEmitter, Host, HostBinding, HostListener, Inject, Injectable, Injector, Input, NgModule, NgZone, Optional, Output, Renderer2, ViewChild, ViewChildren, ViewEncapsulation, animate, forwardRef, state, style, transition, trigger } from '@angular/core';
import { COMPOSITION_BUFFER_MODE, DefaultValueAccessor, FormsModule, NG_VALUE_ACCESSOR, NgControl, NgForm } from '@angular/forms';
import { DOCUMENT, DomSanitizer, ɵgetDOM } from '@angular/platform-browser';
import { Subject } from 'rxjs/Subject';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
import { CommonModule } from '@angular/common';
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIFormControl = (function (_super) {
    tslib_1.__extends(WeUIFormControl, _super);
    /**
     * @param {?} __renderer
     * @param {?} __elementRef
     * @param {?} __compositionMode
     */
    function WeUIFormControl(__renderer, __elementRef, __compositionMode) {
        var _this = _super.call(this, __renderer, __elementRef, __compositionMode) || this;
        _this.__renderer = __renderer;
        _this.__elementRef = __elementRef;
        _this.__compositionMode = __compositionMode;
        /**
         * 是否禁用
         */
        _this.disabled = false;
        /**
         * The internal data model
         */
        _this._value = '';
        _this._cls_cell = true;
        _this._cls_control = true;
        _this.id = WeUIFormControl.registerControl();
        return _this;
    }
    Object.defineProperty(WeUIFormControl.prototype, "innerValue", {
        /**
         * The value of the input ngModel (model -> view)
         * @return {?}
         */
        get: function () {
            return this._value;
        },
        /**
         * The value of the input ngModel (view -> model)
         * @param {?} newValue
         * @return {?}
         */
        set: function (newValue) {
            if (this._value !== newValue) {
                this._value = newValue;
                // view -> model -> outside world (ie. NgModel on this control)
                this.onChange(newValue);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 注册控件
     * @return {?}
     */
    WeUIFormControl.registerControl = function () {
        return 'weui-control-' + (++WeUIFormControl.count).toString();
    };
    /**
     * Write a new value to the element. (From ControlValueAccessor interface)
     * @param {?} value
     * @return {?}
     */
    WeUIFormControl.prototype.writeValue = function (value) {
        this._value = value;
    };
    /**
     * 设置禁用状态 (From ControlValueAccessor interface)
     * @param {?} isDisabled
     * @return {?}
     */
    WeUIFormControl.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    return WeUIFormControl;
}(DefaultValueAccessor));
/**
 * 计数器
 */
WeUIFormControl.count = 0;
WeUIFormControl.decorators = [
    { type: Component, args: [{
                selector: 'WeUIFormControl',
                template: ""
            },] },
];
/**
 * @nocollapse
 */
WeUIFormControl.ctorParameters = function () { return [
    { type: Renderer2, decorators: [{ type: Inject, args: [Renderer2,] },] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] },] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUIFormControl.propDecorators = {
    'id': [{ type: Input },],
    'name': [{ type: Input },],
    'label': [{ type: Input },],
    'disabled': [{ type: Input },],
    'value': [{ type: Input },],
    '_cls_cell': [{ type: HostBinding, args: ['class.weui-cell',] },],
    '_cls_control': [{ type: HostBinding, args: ['class.weui-control',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 徽章
 */
var WeUIBadge = (function () {
    function WeUIBadge() {
        /**
         * 颜色，取值：default、primary、warn等。默认为default。<br>
         * 自定义的颜色名称与色值，可以定义在 工程根目录/src/theme/variables.scss 文件中的 $colors 对象。
         */
        this.color = 'default';
    }
    return WeUIBadge;
}());
WeUIBadge.decorators = [
    { type: Component, args: [{
                selector: 'weui-badge',
                template: "\n        <span class=\"weui-badge weui-badge_{{color}}\"\n             [ngClass]=\"{'weui-badge_dot': dot !== undefined && dot !== null}\"><ng-content></ng-content></span>"
            },] },
];
/**
 * @nocollapse
 */
WeUIBadge.ctorParameters = function () { return []; };
WeUIBadge.propDecorators = {
    'color': [{ type: Input },],
    'dot': [{ type: Input },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIButton = (function () {
    function WeUIButton() {
        /**
         * 颜色，取值：default、primary、warn等。默认为default。<br>
         * 自定义的颜色名称与色值，可以定义在 工程根目录/src/theme/variables.scss 文件中的 $colors 对象。
         */
        this.color = 'default';
        /**
         * 正在加载
         */
        this.loading = false;
        /**
         * 禁用样式
         */
        this.disabled = false;
    }
    /**
     * @return {?}
     */
    WeUIButton.prototype.getColorStyle = function () {
        return 'weui-btn_' + (this.plain !== undefined ? 'plain-' : '') + this.color;
    };
    return WeUIButton;
}());
WeUIButton.decorators = [
    { type: Component, args: [{
                selector: 'weui-button',
                template: "\n        <a href=\"javascript:;\" class=\"weui-btn {{getColorStyle()}}\"\n            [ngClass]=\"{\n                'weui-btn_disabled': disabled && plain == undefined,\n                'weui-btn_plain-disabled': disabled && plain !== undefined,\n                'weui-btn_loading': loading,\n                'weui-btn_mini': mini !== undefined\n            }\">\n            <i class=\"weui-loading\" *ngIf=\"loading\"></i>\n            <ng-content></ng-content>\n        </a>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIButton.ctorParameters = function () { return []; };
WeUIButton.propDecorators = {
    'color': [{ type: Input },],
    'plain': [{ type: Input },],
    'mini': [{ type: Input },],
    'loading': [{ type: Input },],
    'disabled': [{ type: Input },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUISubmit = (function () {
    function WeUISubmit() {
        /**
         * 颜色，取值：default、primary、warn等。默认为default。<br>
         * 自定义的颜色名称与色值，可以定义在 工程根目录/src/theme/variables.scss 文件中的 $colors 对象。
         */
        this.color = 'default';
        /**
         * 按钮类型
         */
        this.type = 'submit';
        /**
         * 正在加载
         */
        this.loading = false;
        /**
         * 禁用样式
         */
        this.disabled = false;
    }
    /**
     * @return {?}
     */
    WeUISubmit.prototype.getColorStyle = function () {
        return 'weui-btn_' + (this.plain !== undefined ? 'plain-' : '') + this.color;
    };
    return WeUISubmit;
}());
WeUISubmit.decorators = [
    { type: Component, args: [{
                selector: 'weui-submit',
                template: "\n        <button [attr.type]=\"type\" class=\"weui-btn {{getColorStyle()}}\"\n            [ngClass]=\"{\n                'weui-btn_disabled': disabled && plain == undefined,\n                'weui-btn_plain-disabled': disabled && plain !== undefined,\n                'weui-btn_loading': loading,\n                'weui-btn_mini': mini !== undefined\n            }\">\n            <i class=\"weui-loading\" *ngIf=\"loading\"></i>\n            <ng-content></ng-content>\n        </button>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUISubmit.ctorParameters = function () { return []; };
WeUISubmit.propDecorators = {
    'color': [{ type: Input },],
    'type': [{ type: Input },],
    'plain': [{ type: Input },],
    'mini': [{ type: Input },],
    'loading': [{ type: Input },],
    'disabled': [{ type: Input },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WEUI_FORM_CONTROL_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return WeUICheckbox; }),
    multi: true
};
var WeUICheckbox = (function (_super) {
    tslib_1.__extends(WeUICheckbox, _super);
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} compositionMode
     */
    function WeUICheckbox(renderer, elementRef, compositionMode) {
        var _this = _super.call(this, renderer, elementRef, compositionMode) || this;
        _this.renderer = renderer;
        _this.elementRef = elementRef;
        _this.compositionMode = compositionMode;
        /**
         * 是否已选中
         */
        _this.checked = false;
        /**
         * 选中的值
         */
        _this.values = [];
        /**
         * 扩展样式
         */
        _this._cls_check_label = true;
        _this.value = 'on'; // default value
        return _this;
    }
    Object.defineProperty(WeUICheckbox.prototype, "innerValue", {
        /**
         * The value of the input ngModel。 (view -> model)
         * @param {?} checked
         * @return {?}
         */
        set: function (checked) {
            this._value = checked || false;
            var /** @type {?} */ index = this.values.indexOf(this.value);
            if (checked) {
                if (index === -1) {
                    this.values.push(this.value);
                }
            }
            else {
                if (index >= 0) {
                    this.values.splice(index, 1);
                }
            }
            // view -> model -> outside world (ie. NgModel on this control)
            this.onChange(this.values);
            // console.log('innerValue: id=' + this.id + ', name=' + this.name
            //  + ', values=' + JSON.stringify(this.values) + ', checked=' + this._value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Write a new value to the element. (From ControlValueAccessor interface)
     * @param {?} value
     * @return {?}
     */
    WeUICheckbox.prototype.writeValue = function (value) {
        if (value !== null) {
            this.values = value;
        }
        this.checked = this.values.indexOf(this.value) >= 0;
        _super.prototype.writeValue.call(this, this.checked);
        // console.log('writeValue: id=' + this.id + ', name=' + this.name
        //  + ', values=' + JSON.stringify(this.values) + ', checked=' + this._value);
    };
    return WeUICheckbox;
}(WeUIFormControl));
WeUICheckbox.decorators = [
    { type: Component, args: [{
                selector: 'weui-checkbox',
                providers: [WEUI_FORM_CONTROL_VALUE_ACCESSOR],
                template: "\n        <label class=\"weui-check__label\" [for]=\"id\" (click)=\"onTouched()\">\n            <div class=\"weui-cell__hd\">\n                <input type=\"checkbox\" class=\"weui-check\"\n                    [attr.id]=\"id\" [attr.name]=\"name\" [value]=\"value\"\n                    [checked]=\"checked\" [(ngModel)]=\"innerValue\">\n                <i class=\"weui-icon-checked\"></i>\n            </div>\n            <div class=\"weui-cell__bd\">\n                {{label}}<ng-content></ng-content>\n            </div>\n        </label>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUICheckbox.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUICheckbox.propDecorators = {
    '_cls_check_label': [{ type: HostBinding, args: ['class.weui-check__label',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 菜单选项
 */
/**
 * 多列选择器
 * 参考了 {\@link https://github.com/weui/weui.js/blob/master/src/picker/picker.js}的实现
 */
var WeUIPicker = (function () {
    function WeUIPicker() {
        /**
         * \@i18n
         */
        this.defaults = {
            cancelText: '取消',
            confirmText: '确定'
        };
        /**
         * 多列菜单列表，内容任意，其中<code>label</code>用于显示，<code>value</code>表示值，<code>disabled</code>表示不可选择项。<br>
         * 注意：列的个数不宜超过3个。
         */
        this.menus = [];
        /**
         * 当前选择的值（每列对应一个值）
         */
        this.value = [];
        /**
         * \@i18n 取消
         */
        this.cancelText = this.defaults.cancelText;
        /**
         * \@i18n 确定
         */
        this.confirmText = this.defaults.confirmText;
        /**
         * 关闭事件
         */
        this.close = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
        /**
         * 已显示否
         */
        this.hidden = true;
    }
    Object.defineProperty(WeUIPicker.prototype, "menu", {
        /**
         * 单列菜单列表，内容任意，其中<code>label</code>用于显示，<code>value</code>表示值，<code>disabled</code>表示不可选择项。
         * @param {?} menu
         * @return {?}
         */
        set: function (menu) {
            if (menu !== undefined && menu !== null && menu.length) {
                this.menus = [menu];
            }
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(WeUIPicker.prototype, "hideCls", {
        /**
         * 用于控制控件的可视化
         * @return {?}
         */
        get: function () {
            return this.hidden;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    WeUIPicker.prototype.onAnimationEnd = function (event) {
        if (event.animationName === 'slideDown') {
            this.hidden = true;
            this.close.emit();
        }
    };
    /**
     * 显示菜单
     * @return {?}
     */
    WeUIPicker.prototype.show = function () {
        var _this = this;
        this.hidden = false;
        this.shown = true;
        return new Promise(function (resolve, reject) {
            _this.resolve = resolve;
        });
    };
    /**
     * 隐藏菜单
     * @return {?}
     */
    WeUIPicker.prototype.hide = function () {
        this.shown = false;
    };
    /**
     * 某列数据修改时触发
     *
     * @param {?} option 该列数据的选择项
     * @param {?} index  该列在组内的索引
     * @return {?}
     */
    WeUIPicker.prototype.onChange = function (option, index) {
        if (option) {
            this.value[index] = option;
        }
    };
    /**
     * 点击【确定】按钮时，通过Promise.resolve()输出最终值
     * @return {?}
     */
    WeUIPicker.prototype.onConfirm = function () {
        this.resolve(this.value);
        this.hide();
    };
    /**
     * 点击【取消】按钮时，关闭Picker
     * @return {?}
     */
    WeUIPicker.prototype.onCancel = function () {
        this.hide();
    };
    return WeUIPicker;
}());
WeUIPicker.decorators = [
    { type: Component, args: [{
                selector: 'weui-picker',
                template: "\n        <div class=\"weui-mask\" (click)=\"onCancel()\"\n            [ngClass]=\"{'weui-animate-fade-in': shown, 'weui-animate-fade-out': !shown}\"></div>\n        <div class=\"weui-picker\"\n            [ngClass]=\"{'weui-animate-slide-up': shown, 'weui-animate-slide-down': !shown}\"\n            (animationend)=\"onAnimationEnd($event)\">\n            <div class=\"weui-picker__hd\">\n                <a href=\"javascript:;\" class=\"weui-picker__action\" (click)=\"onCancel()\">{{cancelText}}  </a>\n                <a href=\"javascript:;\" class=\"weui-picker__action\" (click)=\"onConfirm()\">{{confirmText}}</a>\n            </div>\n            <div class=\"weui-picker__bd\">\n                <weui-picker-group tappable class=\"weui-picker__group\"\n                    *ngFor=\"let options of menus; let i = index;\"\n                    [options]=\"options\" [value]=\"value[i]\" (change)=\"onChange($event, i)\"></weui-picker-group>\n            </div>\n        </div>\n    ",
                exportAs: 'weuiPicker'
            },] },
];
/**
 * @nocollapse
 */
WeUIPicker.ctorParameters = function () { return []; };
WeUIPicker.propDecorators = {
    'menu': [{ type: Input },],
    'menus': [{ type: Input },],
    'value': [{ type: Input },],
    'cancelText': [{ type: Input },],
    'confirmText': [{ type: Input },],
    'hideCls': [{ type: HostBinding, args: ['class.weui-hide',] },],
    'close': [{ type: Output },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 获取Window窗口的高度
 */
var getWindowHeight = function () {
    return window.innerHeight;
};
var WeUIPickerGroup = (function () {
    function WeUIPickerGroup() {
        /**
         * 默认参数
         */
        this.defaults = {
            offset: 3,
            rowHeight: 34,
            bodyHeight: 7 * 34,
            inertiaTime: 150,
            slideDuration: 300 // 惯性滑动的动画时间，表现为最终可视化的效果
        };
        /**
         * 当前列表组的选项
         */
        this.options = [];
        /**
         * 修改事件
         */
        this.change = new EventEmitter();
        /**
         * 选项滚动的距离，用于设置translate3d的Y值
         */
        this.distance = 0;
        /**
         * 选项惯性滚动的时间
         */
        this.duration = 0;
    }
    Object.defineProperty(WeUIPickerGroup.prototype, "translate3d", {
        /**
         * @return {?}
         */
        get: function () {
            return "translate3d(0, " + this.distance + "px, 0)";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WeUIPickerGroup.prototype, "transition", {
        /**
         * @return {?}
         */
        get: function () {
            return "all " + this.duration + "ms";
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WeUIPickerGroup.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.init();
        }, 1);
    };
    /**
     * @return {?}
     */
    WeUIPickerGroup.prototype.forceChange = function () {
        this.init();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    WeUIPickerGroup.prototype._start = function (event) {
        this.startY = event.changedTouches[0].pageY;
        this.startTime = +new Date();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    WeUIPickerGroup.prototype._move = function (event) {
        var /** @type {?} */ endTime = +new Date();
        this.endY = event.changedTouches[0].pageY;
        // 计算滑动的速度: 距离 / 时间
        var /** @type {?} */ _distance = this.endY - (this.startY || 0);
        this.speed = _distance / (endTime - this.startTime);
        // 重新设置开始时间、开始位置
        this.startTime = endTime;
        this.startY = this.endY;
        this.duration = 0; // ms
        this.distance += _distance; // 内容移动的距离
        // console.log(`速度=${this.speed} px/ms`);
        event.preventDefault();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    WeUIPickerGroup.prototype._end = function (event) {
        if (!this.startY) {
            return;
        }
        this.endY = event.changedTouches[0].pageY;
        /**
         * 思路:
         * 根据最后一次touchmove事件的速度(speed)，判断是否执行惯性滑动；
         * 如果speed大于1、小于5，则根据速度乘以惯性滑动的时间(如150ms)，计算出应该滑动的距离；
         * 如果speed大于5，则按照屏幕的高度(减去列表高度的一半)，作为该滑动的距离。
         */
        var _speed = Math.abs(this.speed);
        if (_speed >= 5) {
            var /** @type {?} */ windowY = getWindowHeight() - (this.defaults.bodyHeight / 2);
            this.stop(windowY - this.endY);
        }
        else if (_speed >= 1) {
            var /** @type {?} */ diff = this.speed * this.defaults.inertiaTime; // 滑行 150ms,这里直接影响“灵敏度”
            this.stop(diff);
        }
        else {
            this.stop(0);
        }
        this.startY = null;
    };
    /**
     * @return {?}
     */
    WeUIPickerGroup.prototype.init = function () {
        var /** @type {?} */ index = 0;
        if (this.value) {
            var /** @type {?} */ len = this.options && this.options.length || 0;
            for (; index < len; index++) {
                var /** @type {?} */ option = this.options[index];
                if (this.value === option.value || this.value['value'] === option.value) {
                    this.onChange(option, index);
                    this.distance = (this.defaults.offset - index) * this.defaults.rowHeight;
                    return; // 已找到匹配选项，直接返回
                }
            }
            console.warn('Picker has not match defaultValue:', this.value);
        }
        // 没有传入value，或者 有value但是没有匹配的选项
        index = this._getDefaultIndex(this.options);
        this.onChange(this.options[index], index);
        this.distance = this._getDefaultTranslate(this.defaults.offset, this.defaults.rowHeight, this.options);
    };
    /**
     * 当touch事件结束时，根据惯性滑动计算当前在哪个选项，以及列表滚动的最终距离
     *
     * @param {?} diff 惯性滑动的距离
     * @return {?}
     */
    WeUIPickerGroup.prototype.stop = function (diff) {
        var /** @type {?} */ dist = this.distance + diff;
        // 移动到最接近的那一行
        dist = Math.round(dist / this.defaults.rowHeight) * this.defaults.rowHeight;
        var /** @type {?} */ max = this._getMax(this.defaults.offset, this.defaults.rowHeight);
        var /** @type {?} */ min = this._getMin(this.defaults.offset, this.defaults.rowHeight, this.options.length);
        // console.log(`移动距离: dist=${dist}px`);
        // 不要超过最大值或者最小值
        dist = Math.max(Math.min(dist, max), min);
        // 如果是 disabled 的就跳过
        var /** @type {?} */ index = this.defaults.offset - dist / this.defaults.rowHeight;
        while (!!this.options[index] && this.options[index].disabled) {
            diff > 0 ? ++index : --index;
        }
        dist = (this.defaults.offset - index) * this.defaults.rowHeight;
        this.duration = this.defaults.slideDuration; // ms
        this.distance = dist; // px
        // 触发选择事件
        this.onChange(this.options[index], index);
    };
    /**
     * 选择某个选项时触发
     *
     * @param {?} option 选中的选项
     * @param {?=} index  该选项在列表中的索引
     * @return {?}
     */
    WeUIPickerGroup.prototype.onChange = function (option, index) {
        this.change.emit(option);
    };
    /**
     * get max translate
     *
     * @param {?} offset
     * @param {?} rowHeight
     * @return {?}
     */
    WeUIPickerGroup.prototype._getMax = function (offset, rowHeight) {
        return offset * rowHeight;
    };
    /**
     * get min translate
     *
     * @param {?} offset
     * @param {?} rowHeight
     * @param {?} length
     * @return {?}
     */
    WeUIPickerGroup.prototype._getMin = function (offset, rowHeight, length) {
        return -(rowHeight * (length - offset - 1));
    };
    /**
     * get index of middle item
     *
     * @param {?} options
     * @return {?}
     */
    WeUIPickerGroup.prototype._getDefaultIndex = function (options) {
        var /** @type {?} */ current = Math.floor(options.length / 2);
        var /** @type {?} */ count = 0;
        while (!!options[current] && options[current].disabled) {
            current = ++current % options.length;
            count++;
            if (count > options.length) {
                throw new Error('No selectable item.');
            }
        }
        return current;
    };
    /**
     * @param {?} offset
     * @param {?} rowHeight
     * @param {?} options
     * @return {?}
     */
    WeUIPickerGroup.prototype._getDefaultTranslate = function (offset, rowHeight, options) {
        var /** @type {?} */ currentIndex = this._getDefaultIndex(options);
        return (offset - currentIndex) * rowHeight;
    };
    return WeUIPickerGroup;
}());
WeUIPickerGroup.decorators = [
    { type: Component, args: [{
                selector: 'weui-picker-group',
                template: "\n        <div class=\"weui-picker__mask\"></div>\n        <div class=\"weui-picker__indicator\"></div>\n        <div class=\"weui-picker__content\" [ngStyle]=\"{\n                '-webkit-transform': translate3d, 'transform': translate3d,\n                '-webkit-transition': transition, 'transition': transition\n            }\">\n            <div class=\"weui-picker__item\" *ngFor=\"let option of options\"\n                [ngClass]=\"{'weui-picker__item_disabled': option.disabled}\">{{option.label}}</div>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIPickerGroup.ctorParameters = function () { return []; };
WeUIPickerGroup.propDecorators = {
    'options': [{ type: Input },],
    'value': [{ type: Input },],
    'change': [{ type: Output },],
    '_start': [{ type: HostListener, args: ['touchstart', ['$event'],] },],
    '_move': [{ type: HostListener, args: ['touchmove', ['$event'],] },],
    '_end': [{ type: HostListener, args: ['touchend', ['$event'],] },],
};
/**
 * 支持的格式：(1) 纯数字；(2)范围: min-max；(3)叠加：mum/step
 */
var regex = /^(\d+)(?:-(\d+))?(?:\/(\d+))?$/g;
/**
 * 日[1-31]、月[1-12]、周[0-6]（周日-周六）的取值范围
 */
var constraints = [[1, 31], [1, 12], [0, 6]];
/**
 * Schedule
 */
var Schedule = (function () {
    /**
     * 构造
     * @param {?} fields [[日..], [月..], [周..]]
     * @param {?} _start 起始日期
     * @param {?} _end   终止日期
     */
    function Schedule(fields, _start, _end) {
        this.fields = fields;
        this._start = _start;
        this._end = _end;
        /**
         * dayOfMonth
         * @type {Array}
         */
        this._dates = fields[0];
        /**
         * month
         * @type {Array}
         */
        this._months = fields[1];
        /**
         * dayOfWeek
         * @type {Array}
         */
        this._days = fields[2];
        /**
         * cursor
         * @type {Date}
         * @private
         */
        this._pointer = _start;
    }
    /**
     * @return {?}
     */
    Schedule.prototype._findNext = function () {
        var /** @type {?} */ next;
        while (true) {
            if (this._end.getTime() - this._pointer.getTime() < 0) {
                throw new Error("out of range, end is " + this._end + ", current is " + this._pointer);
            }
            var /** @type {?} */ month = this._pointer.getMonth();
            var /** @type {?} */ date = this._pointer.getDate();
            var /** @type {?} */ day = this._pointer.getDay();
            if (this._months.indexOf(month + 1) === -1) {
                this._pointer.setMonth(month + 1);
                this._pointer.setDate(1);
                continue;
            }
            if (this._dates.indexOf(date) === -1) {
                this._pointer.setDate(date + 1);
                continue;
            }
            if (this._days.indexOf(day) === -1) {
                this._pointer.setDate(date + 1);
                continue;
            }
            next = new Date(this._pointer.getTime());
            break;
        }
        return next;
    };
    /**
     * fetch next data
     * @return {?}
     */
    Schedule.prototype.next = function () {
        var /** @type {?} */ value = this._findNext();
        // move next date
        this._pointer.setDate(this._pointer.getDate() + 1);
        return {
            value: value,
            done: !this.hasNext()
        };
    };
    /**
     * has next
     * @return {?}
     */
    Schedule.prototype.hasNext = function () {
        try {
            this._findNext();
            return true;
        }
        catch (e) {
            return false;
        }
    };
    return Schedule;
}());
/**
 * 解析表达式，并获得合法的数字的列表，如合法的月份数
 *
 * @param {?} field 日期cron表达式中通过空格分割后的某一部分(日、月、周)的内容
 * @param {?} _constraints 取值约束，如日期为[1,31]
 * @return {?} number[]
 */
function parseField(field, _constraints) {
    var /** @type {?} */ low = _constraints[0];
    var /** @type {?} */ high = _constraints[1];
    var /** @type {?} */ result = [];
    // * 号等于最低到最高
    field = field.replace(/\*/g, low + '-' + high);
    // 处理 1,2,5-9 这种情况
    var /** @type {?} */ fields = field.split(',');
    var /** @type {?} */ len = fields.length;
    for (var /** @type {?} */ i = 0; i < len; i++) {
        var /** @type {?} */ f = fields[i];
        if (f.match(regex)) {
            f.replace(regex, function ($0, _lower, _upper, _step) {
                // ref to `cron-parser`
                var /** @type {?} */ step = parseInt(_step, 10) || 1;
                // Positive integer higher than constraints[0]
                var /** @type {?} */ lower = Math.min(Math.max(low, Math.abs(+_lower)), high);
                // Positive integer lower than constraints[1]
                var /** @type {?} */ upper = _upper ? Math.min(high, Math.abs(+_upper)) : high;
                // Count from the lower barrier to the upper
                var /** @type {?} */ pointer = lower;
                do {
                    result.push(pointer);
                    pointer += step;
                } while (pointer <= upper);
                return '';
            });
        }
    }
    return result;
}
/**
 * 解析表达式，并返回Schedule对象
 *
 * @param {?} expr  cron表达式，形如: 日期 月份 周。例如：* * * 或 1,2,5-9 1-6 0-6 或 1/2 * * 等
 * @param {?} start 开始日期
 * @param {?} end   介绍日期
 * @return {?}
 */
function parse(expr, start, end) {
    var /** @type {?} */ atoms = expr.replace(/^\s\s*|\s\s*$/g, '').split(/\s+/);
    var /** @type {?} */ fields = [];
    atoms.forEach(function (atom, index) {
        var /** @type {?} */ constraint = constraints[index];
        fields.push(parseField(atom, constraint));
    });
    return new Schedule(fields, start, end);
}
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 求下一轮的数值
 *
 * @param {?} expr cron 表达式
 * @param {?=} year 年份
 * @param {?=} month 月份，取值范围为：1-12
 * @return {?}
 */
function nexts(expr, year, month) {
    var /** @type {?} */ result = [];
    var /** @type {?} */ atoms = expr.replace(/^\s\s*|\s\s*$/g, '').split(/\s+/);
    // (1) 根据“年” 求可用的“月”
    if (!month) {
        var /** @type {?} */ months = parseField(atoms[1], [1, 12]);
        for (var /** @type {?} */ m = 1; m <= 12; m++) {
            result.push({
                label: m + '月',
                value: m,
                disabled: months.indexOf(m) === -1
            });
        }
        return result;
    }
    // (2) 根据“年、月” 求可用的“日”
    var /** @type {?} */ today = new Date();
    var /** @type {?} */ _year = year || today.getFullYear();
    var /** @type {?} */ maxDate = getActualMaximumDate(_year, month);
    var /** @type {?} */ startDt = new Date(_year, month - 1, 1);
    var /** @type {?} */ endDt = new Date(_year, month - 1, maxDate);
    var /** @type {?} */ interval = parse(expr, startDt, endDt);
    var /** @type {?} */ available = [];
    var /** @type {?} */ obj;
    do {
        obj = interval.next();
        available.push(obj.value.getDate());
    } while (!obj.done);
    for (var /** @type {?} */ i = 1; i <= maxDate; i++) {
        result.push({
            label: i + '日',
            value: i,
            disabled: available.indexOf(i) === -1
        });
    }
    return result;
}
/**
 * 根据年、月求当月的最大日期
 *
 * @param {?} year 年份
 * @param {?} month 月份，取值：1-12（1月-12月）
 * @return {?}
 */
function getActualMaximumDate(year, month) {
    var /** @type {?} */ calendar = new Date(year, month - 1, 31);
    while (calendar.getMonth() !== month - 1) {
        calendar.setDate(calendar.getDate() - 1);
    }
    return calendar.getDate();
}
/**
 * 日期选择器
 */
var WeUIDatePicker = (function (_super) {
    tslib_1.__extends(WeUIDatePicker, _super);
    function WeUIDatePicker() {
        var _this = _super.call(this) || this;
        /**
         * cron 表达式，三位，分别是 dayOfMonth[1-31]，month[1-12] 和 dayOfWeek[0-6]（周日-周六）。<br>
         * 格式: 日期 月份 周。例如：* * * 或 1,2,5-9 1-6 0-6 或 1/2 * * 等。默认：* * *。
         */
        _this.cron = '* * *';
        return _this;
    }
    Object.defineProperty(WeUIDatePicker.prototype, "start", {
        /**
         * 开始日期/年份。取值：<ul>
         * <li>(1)字符串：符合ISO 8601日期格式的字符串，如：2013-01-18 或者 20130118；</li>
         * <li>(2)数值：表示年份，如：2011，则默认加"-01-01"在年份之后 </li>
         * <li>(3)Date：JavaScript Date对象 </li>
         * </ul>
         * @param {?} start
         * @return {?}
         */
        set: function (start) {
            this._startDate = this._parseDate(start, '-01-01');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WeUIDatePicker.prototype, "end", {
        /**
         * 结束日期/年份。取值：<ul>
         * <li>(1)字符串：符合ISO 8601日期格式的字符串，如：2013-01-18 或者 20130118；</li>
         * <li>(2)数值：表示年份，如：2011，则默认加"-12-31"在年份之后 </li>
         * <li>(3)Date：JavaScript Date对象 </li>
         * </ul>
         * @param {?} end
         * @return {?}
         */
        set: function (end) {
            this._endDate = this._parseDate(end, '-12-31');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @param {?} sub
     * @return {?}
     */
    WeUIDatePicker.prototype._parseDate = function (value, sub) {
        if (typeof value === 'string') {
            if (value.length === 4) {
                value = value + sub;
            }
            if (/^(\d{4})-(\d{2})-(\d{2})$/.test(value) || /^(\d{4})(\d{2})(\d{2})$/.test(value)) {
                return new Date(+RegExp.$1, +RegExp.$2 - 1, +RegExp.$3);
            }
        }
        if (typeof value === 'number') {
            return new Date(value + sub);
        }
        return value; // 默认为Date对象
    };
    /**
     * @return {?}
     */
    WeUIDatePicker.prototype.ngOnInit = function () {
        // 设置初始值
        var /** @type {?} */ now = new Date();
        var /** @type {?} */ thisYear = now.getFullYear();
        if (!this._startDate) {
            this.start = thisYear - 10;
        }
        if (!this._endDate) {
            this.end = thisYear + 10;
        }
        if (this._startDate.getTime() > this._endDate.getTime()) {
            _a = [this._endDate, this._startDate], this._startDate = _a[0], this._endDate = _a[1];
        }
        // 年份列表
        var /** @type {?} */ years = [];
        var /** @type {?} */ endYear = this._endDate.getFullYear();
        for (var /** @type {?} */ year = this._startDate.getFullYear(); year <= endYear; year++) {
            years.push({
                label: year + '年',
                value: year
            });
        }
        // 渲染数据列表
        this.menus = [years, [], []];
        var _a;
    };
    /**
     * @return {?}
     */
    WeUIDatePicker.prototype.ngAfterViewInit = function () {
        var /** @type {?} */ monthPicker = this._groups.find(function (group, index) {
            return index === 1;
        });
        var /** @type {?} */ datePicker = this._groups.find(function (group, index) {
            return index === 2;
        });
        if (monthPicker) {
            this.monthPicker = monthPicker;
        }
        if (datePicker) {
            this.datePicker = datePicker;
        }
        // 更新月份、日期列表
        this._updateMonthPickerList();
    };
    /**
     * 选择年或月，触发子列表的变更
     *
     * @param {?} option 被选择的选项
     * @param {?} index 哪个列（索引）
     * @return {?}
     */
    WeUIDatePicker.prototype.onChange = function (option, index) {
        if (!option) {
            return; // 未选择任何值，不触发子列表的变更
        }
        _super.prototype.onChange.call(this, option, index);
        // 变更“年份”时，修改月份列表；如果月份的值为2月，则也修改日期列表
        if (index === 0) {
            this._updateMonthPickerList();
        }
        // 变更“月份”时，修改日期列表
        if (index === 1) {
            this._updateDatePickerList();
        }
    };
    /**
     * @return {?}
     */
    WeUIDatePicker.prototype._updateMonthPickerList = function () {
        if (!this.monthPicker) {
            return;
        }
        if (!this.value || !this.value.length || !this.value[0]) {
            return;
        }
        var /** @type {?} */ year = this.value[0].value;
        var /** @type {?} */ months = nexts(this.cron, year);
        var /** @type {?} */ options = this.monthPicker.options;
        if (months.length !== options.length) {
            this.monthPicker.options = months;
            this.monthPicker.forceChange();
        }
        // 强制日期列表更新
        this._updateDatePickerList();
    };
    /**
     * @return {?}
     */
    WeUIDatePicker.prototype._updateDatePickerList = function () {
        if (!this.datePicker) {
            return;
        }
        var /** @type {?} */ year = this.value[0].value;
        var /** @type {?} */ month = this.value[1].value;
        var /** @type {?} */ dates = nexts(this.cron, year, month);
        this.value[2] = null; // 清空值
        this.datePicker.options = dates;
        this.datePicker.forceChange();
    };
    return WeUIDatePicker;
}(WeUIPicker));
WeUIDatePicker.decorators = [
    { type: Component, args: [{
                selector: 'weui-date-picker',
                template: "\n        <div class=\"weui-mask\" (click)=\"onCancel()\"\n            [ngClass]=\"{'weui-animate-fade-in': shown, 'weui-animate-fade-out': !shown}\"></div>\n        <div class=\"weui-picker\"\n            [ngClass]=\"{'weui-animate-slide-up': shown, 'weui-animate-slide-down': !shown}\"\n            (animationend)=\"onAnimationEnd($event)\">\n            <div class=\"weui-picker__hd\">\n                <a href=\"javascript:;\" class=\"weui-picker__action\" (click)=\"onCancel()\">{{cancelText}}  </a>\n                <a href=\"javascript:;\" class=\"weui-picker__action\" (click)=\"onConfirm()\">{{confirmText}}</a>\n            </div>\n            <div class=\"weui-picker__bd\">\n                <weui-picker-group tappable class=\"weui-picker__group\"\n                    *ngFor=\"let options of menus; let i = index;\"\n                    [options]=\"options\" [value]=\"value[i]\" (change)=\"onChange($event, i)\"></weui-picker-group>\n            </div>\n        </div>\n    ",
                exportAs: 'weuiDatePicker'
            },] },
];
/**
 * @nocollapse
 */
WeUIDatePicker.ctorParameters = function () { return []; };
WeUIDatePicker.propDecorators = {
    'cron': [{ type: Input },],
    'start': [{ type: Input },],
    'end': [{ type: Input },],
    '_groups': [{ type: ViewChildren, args: [WeUIPickerGroup,] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 页脚
 */
var WeUIFooter = (function () {
    function WeUIFooter() {
        /**
         * 固定在底部
         */
        this.fixedAtBottom = false;
        /**
         * 基本样式
         */
        this._cls_footer = true;
    }
    Object.defineProperty(WeUIFooter.prototype, "fixedBottomCls", {
        /**
         * 固定在底部样式
         * @return {?}
         */
        get: function () {
            return this.fixedAtBottom;
        },
        enumerable: true,
        configurable: true
    });
    return WeUIFooter;
}());
WeUIFooter.decorators = [
    { type: Component, args: [{
                selector: 'weui-footer',
                template: '<ng-content></ng-content>'
            },] },
];
/**
 * @nocollapse
 */
WeUIFooter.ctorParameters = function () { return []; };
WeUIFooter.propDecorators = {
    'fixedAtBottom': [{ type: Input },],
    '_cls_footer': [{ type: HostBinding, args: ['class.weui-footer',] },],
    'fixedBottomCls': [{ type: HostBinding, args: ['class.weui-footer_fixed-bottom',] },],
};
var WeUIFooterText = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     * @param {?} _container
     */
    function WeUIFooterText(_renderer, _elementRef, _container) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._container = _container;
        if (_container !== null) {
            _renderer.addClass(_elementRef.nativeElement, 'weui-footer__text');
        }
    }
    return WeUIFooterText;
}());
WeUIFooterText.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'weui-footer-text,[weui-footer-text]'
            },] },
];
/**
 * @nocollapse
 */
WeUIFooterText.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: WeUIFooter, decorators: [{ type: Optional }, { type: Host },] },
]; };
var WeUIFooterLinks = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     * @param {?} _container
     */
    function WeUIFooterLinks(_renderer, _elementRef, _container) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._container = _container;
        if (_container !== null) {
            _renderer.addClass(_elementRef.nativeElement, 'weui-footer__links');
        }
    }
    return WeUIFooterLinks;
}());
WeUIFooterLinks.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'weui-footer-links,[weui-footer-links]'
            },] },
];
/**
 * @nocollapse
 */
WeUIFooterLinks.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: WeUIFooter, decorators: [{ type: Optional }, { type: Host },] },
]; };
var WeUIFooterLink = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     * @param {?} _container
     */
    function WeUIFooterLink(_renderer, _elementRef, _container) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._container = _container;
        if (_container !== null) {
            _renderer.addClass(_elementRef.nativeElement, 'weui-footer__link');
        }
    }
    return WeUIFooterLink;
}());
WeUIFooterLink.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'a'
            },] },
];
/**
 * @nocollapse
 */
WeUIFooterLink.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: WeUIFooterLinks, decorators: [{ type: Optional }, { type: Host },] },
]; };
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIGallery = (function () {
    function WeUIGallery() {
        /**
         * 是否显示删除按钮
         */
        this.canDelete = true;
        /**
         * 删除事件
         */
        this.delete = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
        this._show = false;
    }
    /**
     * 删除照片
     * @param {?} event
     * @return {?}
     */
    WeUIGallery.prototype.deleteImage = function (event) {
        this.delete.emit(this.image);
        event.preventDefault();
    };
    /**
     * 显示图片
     * @return {?}
     */
    WeUIGallery.prototype.show = function () {
        var _this = this;
        this._show = true;
        setTimeout(function () {
            _this.shown = true;
        }, 10);
    };
    /**
     * 隐藏图片
     * @return {?}
     */
    WeUIGallery.prototype.hide = function () {
        var _this = this;
        this.shown = false;
        setTimeout(function () {
            _this._show = false;
        }, 200);
    };
    return WeUIGallery;
}());
WeUIGallery.decorators = [
    { type: Component, args: [{
                selector: 'weui-gallery',
                template: "\n        <div class=\"weui-gallery weui-transition-opacity\" [ngClass]=\"{'weui-block': _show, 'weui-fadeIn': shown}\">\n            <span class=\"weui-gallery__img\" [style.background-image]=\"image?.fileURL\"></span>\n            <div class=\"weui-gallery__opr\" *ngIf=\"canDelete\">\n                <a href=\"javascript:;\" class=\"weui-gallery__del\" (click)=\"deleteImage($event)\">\n                    <i class=\"weui-icon-delete weui-icon_gallery-delete\"></i>\n                </a>\n            </div>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIGallery.ctorParameters = function () { return []; };
WeUIGallery.propDecorators = {
    'image': [{ type: Input },],
    'canDelete': [{ type: Input },],
    'delete': [{ type: Output },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WEUI_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return WeUIInput; }),
    multi: true
};
var WeUIInput = (function (_super) {
    tslib_1.__extends(WeUIInput, _super);
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} compositionMode
     */
    function WeUIInput(renderer, elementRef, compositionMode) {
        var _this = _super.call(this, renderer, elementRef, compositionMode) || this;
        _this.renderer = renderer;
        _this.elementRef = elementRef;
        _this.compositionMode = compositionMode;
        /**
         * 控件类型：text, number, tel, email, password, date, datetime-local等
         */
        _this.type = 'text';
        /**
         * 是否必填
         */
        _this.required = false;
        /**
         * 是否只读
         */
        _this.readonly = false;
        /**
         * autocomplete 属性规定输入字段是否应该启用自动完成功能。默认为on。<br>
         * 自动完成允许浏览器预测对字段的输入。当用户在字段开始键入时，浏览器基于之前键入过的值，应该显示出在字段中填写的选项。
         */
        _this.autocomplete = true;
        /**
         * 是否监听输入长度
         */
        _this.countChars = false;
        /**
         * 输入域无效时显示告警图标，默认为true
         */
        _this.showWarnIcon = true;
        /**
         * 行数，当类型是textarea时
         */
        _this.rows = 3;
        /**
         * \@output {event} Expression to call when the input no longer has focus
         */
        _this.blur = new EventEmitter();
        /**
         * \@output {event} Expression to call when the input has focus
         */
        _this.focus = new EventEmitter();
        /**
         * 扩展样式，如：weui-cell_example
         */
        _this._cls_check_label = true;
        return _this;
    }
    Object.defineProperty(WeUIInput.prototype, "warnCls", {
        /**
         * 扩展样式，如：weui-cell_example
         * @return {?}
         */
        get: function () {
            return this.shouldWarn();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    WeUIInput.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ changed = changes['baseCls'];
        if (changed) {
            var /** @type {?} */ _el = (this.elementRef.nativeElement);
            if (changed.previousValue) {
                _el.classList.remove("weui-cell_" + changed.previousValue);
            }
            if (changed.currentValue) {
                _el.classList.add("weui-cell_" + changed.currentValue);
            }
        }
    };
    /**
     * @return {?}
     */
    WeUIInput.prototype.shouldWarn = function () {
        return this.state.invalid === true &&
            (this.state.touched === true || this.state.dirty === true);
    };
    /**
     * @return {?}
     */
    WeUIInput.prototype.shouldShowWarnIcon = function () {
        return this.showWarnIcon && this.shouldWarn();
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    WeUIInput.prototype.onBlur = function (ev) {
        this.onTouched(); // set your control to 'touched'
        this.blur.emit(ev);
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    WeUIInput.prototype.onFocus = function (ev) {
        this.focus.emit(ev);
    };
    /**
     * 统计字符总长度
     * @return {?}
     */
    WeUIInput.prototype.getCharCount = function () {
        return (this._value && this._value.length) || 0;
    };
    /**
     * 当事件keyup、change时，当字数大于等于maxLength后，禁止输入
     * @param {?} event
     * @return {?}
     */
    WeUIInput.prototype.onKeyboardEvent = function (event) {
        var /** @type {?} */ maxLen = parseInt(this.maxlength, 10);
        if (!isNaN(maxLen) && this.getCharCount() >= maxLen) {
            if (event.keyCode !== 46 && event.keyCode !== 8) {
                event.preventDefault(); // 禁止输入
            }
        }
    };
    return WeUIInput;
}(WeUIFormControl));
WeUIInput.decorators = [
    { type: Component, args: [{
                selector: 'weui-input',
                template: "\n        <div class=\"weui-cell__hd\">\n            {{label}}<ng-content select=\"label\"></ng-content>\n        </div>\n        <div class=\"weui-cell__bd\">\n            <div class=\"weui-input-area\" *ngIf=\"type !== 'textarea'\">\n                <input #input class=\"weui-input\" [type]=\"type\"\n                    [attr.id]=\"id\" [attr.name]=\"name\" [placeholder]=\"placeholder\"\n                    [attr.maxlength]=\"maxlength\" [attr.minlength]=\"minlength\" [pattern]=\"pattern\"\n                    [attr.max]=\"max\" [attr.min]=\"min\" [required]=\"required\" [readonly]=\"readonly\"\n                    [disabled]=\"disabled\" [autocomplete]=\"autocomplete ? 'on' : 'off'\"\n                    [(ngModel)]=\"innerValue\"\n                    (blur)=\"onBlur($event)\" (focus)=\"onFocus($event)\"\n                    (keydown)=\"onKeyboardEvent($event)\" (keyup)=\"onKeyboardEvent($event)\" />\n                <div *ngIf=\"maxlength && countChars\" class=\"weui-text-counter\">\n                    <span>{{getCharCount()}}</span>/{{maxlength}}\n                </div>\n            </div>\n\n            <div class=\"weui-input-area\" *ngIf=\"type == 'textarea'\">\n                <textarea #textarea class=\"weui-textarea\"\n                    [placeholder]=\"placeholder\" [rows]=\"rows\"\n                    [required]=\"required\" [readonly]=\"readonly\"\n                    [attr.minlength]=\"minlength\" [attr.maxlength]=\"maxlength\"\n                    [(ngModel)]=\"innerValue\"\n                    (keydown)=\"onKeyboardEvent($event)\" (keyup)=\"onKeyboardEvent($event)\"></textarea>\n                <div *ngIf=\"maxlength && countChars\" class=\"weui-textarea-counter\">\n                    <span>{{getCharCount()}}</span>/{{maxlength}}\n                </div>\n            </div>\n        </div>\n        <div class=\"weui-cell__ft\">\n            <ng-content select=\"[weui-last]\"></ng-content>\n            <i class=\"weui-icon-warn\" *ngIf=\"shouldShowWarnIcon()\"></i>\n        </div>\n    ",
                providers: [WEUI_INPUT_CONTROL_VALUE_ACCESSOR],
                encapsulation: ViewEncapsulation.None
            },] },
];
/**
 * @nocollapse
 */
WeUIInput.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUIInput.propDecorators = {
    'type': [{ type: Input },],
    'pattern': [{ type: Input },],
    'placeholder': [{ type: Input },],
    'maxlength': [{ type: Input },],
    'minlength': [{ type: Input },],
    'max': [{ type: Input },],
    'min': [{ type: Input },],
    'required': [{ type: Input },],
    'readonly': [{ type: Input },],
    'autocomplete': [{ type: Input },],
    'countChars': [{ type: Input },],
    'baseCls': [{ type: Input },],
    'showWarnIcon': [{ type: Input },],
    'rows': [{ type: Input },],
    'blur': [{ type: Output },],
    'focus': [{ type: Output },],
    'state': [{ type: ContentChild, args: [NgControl,] },],
    '_cls_check_label': [{ type: HostBinding, args: ['class.weui-check__label',] },],
    'warnCls': [{ type: HostBinding, args: ['class.weui-cell_warn',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIItems = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     */
    function WeUIItems(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._cls_cells = true;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    WeUIItems.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ changed = changes['baseCls'];
        if (changed) {
            var /** @type {?} */ _el = (this._elementRef.nativeElement);
            if (changed.previousValue) {
                _el.classList.remove("weui-cells_" + changed.previousValue);
            }
            if (changed.currentValue) {
                _el.classList.add("weui-cells_" + changed.currentValue);
            }
        }
    };
    return WeUIItems;
}());
WeUIItems.decorators = [
    { type: Component, args: [{
                selector: 'weui-items',
                template: "<ng-content></ng-content>"
            },] },
];
/**
 * @nocollapse
 */
WeUIItems.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
]; };
WeUIItems.propDecorators = {
    'baseCls': [{ type: Input },],
    '_cls_cells': [{ type: HostBinding, args: ['class.weui-cells',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIItem = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     */
    function WeUIItem(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._cls_cell = true;
        this._cls_item = true;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    WeUIItem.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ changed = changes['baseCls'];
        if (changed) {
            var /** @type {?} */ _el = (this._elementRef.nativeElement);
            if (changed.previousValue) {
                _el.classList.remove("weui-cell_" + changed.previousValue);
            }
            if (changed.currentValue) {
                _el.classList.add("weui-cell_" + changed.currentValue);
            }
        }
    };
    /**
     * @param {?} cls
     * @return {?}
     */
    WeUIItem.prototype.addClass = function (cls) {
        this._renderer.addClass(this._elementRef.nativeElement, cls);
    };
    return WeUIItem;
}());
WeUIItem.decorators = [
    { type: Component, args: [{
                selector: 'weui-item',
                template: "\n        <div class=\"weui-cell__hd\">\n            <ng-content select=\"[weui-start]\"></ng-content>\n        </div>\n        <div class=\"weui-cell__bd\">\n            <ng-content select=\"[weui-content]\"></ng-content>\n            <ng-content></ng-content>\n        </div>\n        <div class=\"weui-cell__ft\">\n            <ng-content select=\"[weui-last]\"></ng-content>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIItem.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
]; };
WeUIItem.propDecorators = {
    'baseCls': [{ type: Input },],
    '_cls_cell': [{ type: HostBinding, args: ['class.weui-cell',] },],
    '_cls_item': [{ type: HostBinding, args: ['class.weui-item',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUILink = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     * @param {?} sanitizer
     */
    function WeUILink(_renderer, _elementRef, sanitizer) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this.sanitizer = sanitizer;
        /**
         * 不显示最右边的箭头。默认为false，即：显示箭头。
         */
        this.noPushArrow = false;
        this._cls_cell = true;
        this._cls_item = true;
        this._cls_cell_access = true;
    }
    Object.defineProperty(WeUILink.prototype, "href", {
        /**
         * 链接
         * @param {?} _href
         * @return {?}
         */
        set: function (_href) {
            this._href = this.sanitizer.bypassSecurityTrustUrl(_href);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WeUILink.prototype, "_cls_cell_access_noarrow", {
        /**
         * @return {?}
         */
        get: function () {
            return this.noPushArrow;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    WeUILink.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ changed = changes['baseCls'];
        if (changed) {
            var /** @type {?} */ _el = (this._elementRef.nativeElement);
            if (changed.previousValue) {
                _el.classList.remove("weui-cell_" + changed.previousValue);
            }
            if (changed.currentValue) {
                _el.classList.add("weui-cell_" + changed.currentValue);
            }
        }
    };
    return WeUILink;
}());
WeUILink.decorators = [
    { type: Component, args: [{
                selector: 'weui-link',
                template: "\n        <a class=\"weui-cell_access\" [href]=\"_href\">\n            <div class=\"weui-cell__hd\">\n                <ng-content select=\"[weui-start]\"></ng-content>\n            </div>\n            <div class=\"weui-cell__bd\">\n                <ng-content select=\"[weui-content]\"></ng-content>\n                <ng-content></ng-content>\n            </div>\n            <div class=\"weui-cell__ft\">\n                <ng-content select=\"[weui-last]\"></ng-content>\n            </div>\n        </a>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUILink.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: DomSanitizer, },
]; };
WeUILink.propDecorators = {
    'noPushArrow': [{ type: Input },],
    'baseCls': [{ type: Input },],
    'href': [{ type: Input },],
    '_cls_cell': [{ type: HostBinding, args: ['class.weui-cell',] },],
    '_cls_item': [{ type: HostBinding, args: ['class.weui-item',] },],
    '_cls_cell_access': [{ type: HostBinding, args: ['class.weui-cell_access',] },],
    '_cls_cell_access_noarrow': [{ type: HostBinding, args: ['class.weui-cell_access-noarrow',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 加载更多
 */
var WeUILoadmore = (function () {
    function WeUILoadmore() {
        /**
         * 正在加载
         */
        this.loading = false;
    }
    return WeUILoadmore;
}());
WeUILoadmore.decorators = [
    { type: Component, args: [{
                selector: 'weui-loadmore',
                template: "\n        <div class=\"weui-loadmore\" [ngClass]=\"{\n                'weui-loadmore_line': line !== undefined && line !== null,\n                'weui-loadmore_dot': dot !== undefined && dot !== null }\">\n            <i class=\"weui-loading\" *ngIf=\"loading\"></i>\n            <span class=\"weui-loadmore__tips\"><ng-content></ng-content></span>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUILoadmore.ctorParameters = function () { return []; };
WeUILoadmore.propDecorators = {
    'line': [{ type: Input },],
    'dot': [{ type: Input },],
    'loading': [{ type: Input },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUINavBarItem = (function () {
    /**
     * @param {?} navbar
     */
    function WeUINavBarItem(navbar) {
        this.navbar = navbar;
        /**
         * 设置基本样式
         */
        this.barItemCls = true;
        this._activated = false;
    }
    Object.defineProperty(WeUINavBarItem.prototype, "activatedCls", {
        /**
         * 设置激活样式
         * @return {?}
         */
        get: function () {
            return this._activated;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 点击触发激活
     *
     * @param {?} event 点击事件
     * @return {?}
     */
    WeUINavBarItem.prototype.onClick = function (event) {
        if (this.navbar) {
            this.navbar.activate(this);
        }
    };
    /**
     * 由父级控件传入
     *
     * @param {?} activated 激活状态
     * @return {?}
     */
    WeUINavBarItem.prototype.activated = function (activated) {
        this._activated = activated;
    };
    /**
     * @return {?}
     */
    WeUINavBarItem.prototype.isActivated = function () {
        return this._activated;
    };
    return WeUINavBarItem;
}());
WeUINavBarItem.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'weui-navbar-item,[weui-navbar-item]'
            },] },
];
/**
 * @nocollapse
 */
WeUINavBarItem.ctorParameters = function () { return [
    { type: WeUINavBar, decorators: [{ type: Optional }, { type: Host }, { type: Inject, args: [forwardRef(function () { return WeUINavBar; }),] },] },
]; };
WeUINavBarItem.propDecorators = {
    'value': [{ type: Input },],
    'barItemCls': [{ type: HostBinding, args: ['class.weui-navbar__item',] },],
    'activatedCls': [{ type: HostBinding, args: ['class.weui-bar__item_on',] },],
    'onClick': [{ type: HostListener, args: ['click', ['$event'],] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUINavBar = (function () {
    function WeUINavBar() {
        /**
         * 初始激活的子对象
         */
        this.activeIndex = 0;
        /**
         * 设置基本样式
         */
        this.barCls = true;
        /**
         * 激活事件
         */
        this.activated = new EventEmitter();
    }
    /**
     * @return {?}
     */
    WeUINavBar.prototype.ngAfterViewInit = function () {
        this.activateAt(this.activeIndex);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WeUINavBar.prototype.getAt = function (index) {
        return (this.items.find(function (item, i) { return i === index; }));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WeUINavBar.prototype.activate = function (item) {
        this._activated = item;
        this.items.forEach(function (child) {
            child.activated(child === item);
        });
        this.activated.emit(item);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WeUINavBar.prototype.activateAt = function (index) {
        var /** @type {?} */ item = this.getAt(index);
        if (item) {
            this.activate(item);
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WeUINavBar.prototype.isActivated = function (item) {
        return item.isActivated();
    };
    return WeUINavBar;
}());
WeUINavBar.decorators = [
    { type: Component, args: [{
                selector: 'weui-navbar',
                template: "<ng-content></ng-content>"
            },] },
];
/**
 * @nocollapse
 */
WeUINavBar.ctorParameters = function () { return []; };
WeUINavBar.propDecorators = {
    'activeIndex': [{ type: Input },],
    'barCls': [{ type: HostBinding, args: ['class.weui-navbar',] },],
    'items': [{ type: ContentChildren, args: [WeUINavBarItem,] },],
    'activated': [{ type: Output },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIProgressBar = (function () {
    function WeUIProgressBar() {
        /**
         * 颜色，取值：default、primary、warn等。默认为default。<br>
         * 自定义的颜色名称与色值，可以定义在 工程根目录/src/theme/variables.scss 文件中的 $colors 对象。
         */
        this.color = 'default';
        /**
         * Value of the progressbar. Defaults to zero.
         */
        this.value = 0;
        /**
         * 是否可终止 (若为true，则显示关闭按钮)
         */
        this.canTerminate = true;
        /**
         * 终止事件
         */
        this.terminate = new EventEmitter();
    }
    /**
     * @return {?}
     */
    WeUIProgressBar.prototype.onTerminate = function () {
        this.terminate.emit(this);
    };
    return WeUIProgressBar;
}());
WeUIProgressBar.decorators = [
    { type: Component, args: [{
                selector: 'weui-progress-bar',
                template: "\n        <div class=\"weui-progress weui-progress_{{color}}\">\n            <div class=\"weui-progress__bar\">\n                <div class=\"weui-progress__inner-bar\" [style.width]=\"value + '%'\"></div>\n            </div>\n            <a href=\"javascript:;\" class=\"weui-progress__opr\" *ngIf=\"canTerminate\" (click)=\"onTerminate()\">\n                <i class=\"weui-icon-cancel\"></i>\n            </a>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIProgressBar.ctorParameters = function () { return []; };
WeUIProgressBar.propDecorators = {
    'color': [{ type: Input },],
    'value': [{ type: Input },],
    'canTerminate': [{ type: Input },],
    'terminate': [{ type: Output },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WEUI_FORM_CONTROL_VALUE_ACCESSOR$1 = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return WeUIRadio; }),
    multi: true
};
var WeUIRadio = (function (_super) {
    tslib_1.__extends(WeUIRadio, _super);
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} compositionMode
     */
    function WeUIRadio(renderer, elementRef, compositionMode) {
        var _this = _super.call(this, renderer, elementRef, compositionMode) || this;
        _this.renderer = renderer;
        _this.elementRef = elementRef;
        _this.compositionMode = compositionMode;
        /**
         * 基本样式
         */
        _this._cls_check_label = true;
        return _this;
    }
    return WeUIRadio;
}(WeUIFormControl));
WeUIRadio.decorators = [
    { type: Component, args: [{
                selector: 'weui-radio',
                providers: [WEUI_FORM_CONTROL_VALUE_ACCESSOR$1],
                template: "\n        <label class=\"weui-check__label\" [for]=\"id\" (click)=\"onTouched()\">\n            <div class=\"weui-cell__bd\">\n                {{label}}<ng-content></ng-content>\n            </div>\n            <div class=\"weui-cell__ft\">\n                <input type=\"radio\" class=\"weui-check\"\n                    [attr.id]=\"id\" [attr.name]=\"name\" [value]=\"value\"\n                    [disabled]=\"disabled\" [(ngModel)]=\"innerValue\">\n                <span class=\"weui-icon-checked\"></span>\n            </div>\n        </label>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIRadio.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUIRadio.propDecorators = {
    '_cls_check_label': [{ type: HostBinding, args: ['class.weui-check__label',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUISearchBar = (function () {
    function WeUISearchBar() {
        /**
         * \@i18n
         */
        this.defaults = {
            cancelText: '取消',
            searchText: '搜索'
        };
        /**
         * 是否必填
         */
        this.required = true;
        /**
         * \@i18n
         * placeholder 规定帮助用户填写输入字段的提示。
         */
        this.placeholder = this.defaults.searchText;
        /**
         * \@i18n 取消
         */
        this.cancelText = this.defaults.cancelText;
        /**
         * 输入事件
         */
        this.search = new EventEmitter();
        this.focusing = false;
        this.value = '';
        this.searchTerms = new Subject();
    }
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.ngOnInit = function () {
        var _this = this;
        this._searchTermsChangesSubscription = this.searchTerms
            .pipe(debounceTime(300), // wait for 300ms pause in events
        distinctUntilChanged() // ignore if next search term is same as previous
        )
            .subscribe(function (term) {
            _this.search.emit(term);
        });
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.ngOnDestroy = function () {
        if (this._searchTermsChangesSubscription) {
            this._searchTermsChangesSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.doFocus = function () {
        this.searchBox.nativeElement.focus();
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.onFocus = function () {
        this.focusing = true;
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.onBlur = function () {
        if (!this.value.length) {
            this.focusing = false;
        }
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.onCancel = function () {
        this.value = '';
        this.onBlur();
        this.search.emit('');
    };
    /**
     * @param {?} term
     * @return {?}
     */
    WeUISearchBar.prototype.push = function (term) {
        this.searchTerms.next(term);
    };
    /**
     * @return {?}
     */
    WeUISearchBar.prototype.clear = function () {
        this.value = '';
        this.doFocus();
    };
    return WeUISearchBar;
}());
WeUISearchBar.decorators = [
    { type: Component, args: [{
                selector: 'weui-searchbar',
                template: "\n        <div class=\"weui-search-bar\" [ngClass]=\"{'weui-search-bar_focusing': focusing}\">\n            <form class=\"weui-search-bar__form\">\n                <div class=\"weui-search-bar__box\">\n                    <i class=\"weui-icon-search\"></i>\n                    <input #searchBox type=\"search\" name=\"search\" class=\"weui-search-bar__input\"\n                        [placeholder]=\"placeholder\" [required]=\"required\"\n                        (focus)=\"onFocus()\" (blur)=\"onBlur()\" [(ngModel)]=\"value\" (keyup)=\"push(searchBox.value)\" />\n                    <a href=\"javascript:\" class=\"weui-icon-clear\" (click)=\"clear()\"></a>\n                </div>\n                <label class=\"weui-search-bar__label\" (click)=\"doFocus()\">\n                    <i class=\"weui-icon-search\"></i>\n                    <span>{{placeholder}}</span>\n                </label>\n            </form>\n            <a href=\"javascript:\" class=\"weui-search-bar__cancel-btn\" (click)=\"onCancel()\">{{cancelText}}</a>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUISearchBar.ctorParameters = function () { return []; };
WeUISearchBar.propDecorators = {
    'required': [{ type: Input },],
    'placeholder': [{ type: Input },],
    'cancelText': [{ type: Input },],
    'search': [{ type: Output },],
    'searchBox': [{ type: ViewChild, args: ['searchBox',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUISelect = (function () {
    /**
     * @param {?} _renderer
     * @param {?} _elementRef
     * @param {?} _container
     */
    function WeUISelect(_renderer, _elementRef, _container) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this._container = _container;
    }
    /**
     * @return {?}
     */
    WeUISelect.prototype.ngAfterViewInit = function () {
        if (!this._container) {
            return;
        }
        var /** @type {?} */ nativeEl = (this._elementRef.nativeElement);
        this._renderer.addClass(nativeEl, 'weui-select');
        this._container.addClass('weui-cell_select');
        var /** @type {?} */ parentEl = nativeEl.parentElement;
        if (parentEl && parentEl.classList.contains('weui-cell__hd')) {
            this._container.addClass('weui-cell_select-before');
        }
        if (parentEl && parentEl.classList.contains('weui-cell__bd')) {
            this._container.addClass('weui-cell_select-after');
        }
    };
    return WeUISelect;
}());
WeUISelect.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'select'
            },] },
];
/**
 * @nocollapse
 */
WeUISelect.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: WeUIItem, decorators: [{ type: Optional }, { type: Host },] },
]; };
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WEUI_FORM_CONTROL_VALUE_ACCESSOR$2 = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return WeUISlider; }),
    multi: true
};
var WeUISlider = (function (_super) {
    tslib_1.__extends(WeUISlider, _super);
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} compositionMode
     */
    function WeUISlider(renderer, elementRef, compositionMode) {
        var _this = _super.call(this, renderer, elementRef, compositionMode) || this;
        _this.renderer = renderer;
        _this.elementRef = elementRef;
        _this.compositionMode = compositionMode;
        /**
         * 颜色，取值：default、primary、warn等。默认为default。<br>
         * 自定义的颜色名称与色值，可以定义在 工程根目录/src/theme/variables.scss 文件中的 $colors 对象。
         */
        _this.color = 'default';
        /**
         * 显示值（在最后）
         */
        _this.showValue = true;
        /**
         * 禁用样式
         */
        _this.disabled = false;
        /**
         * 最小值
         */
        _this.min = 0;
        /**
         * 最大值
         */
        _this.max = 100;
        /**
         * 取值
         */
        _this.value = 0;
        /**
         * 滑动比例（取值：0-100）
         */
        _this.percent = 0;
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    WeUISlider.prototype.onTouchStart = function (event) {
        if (this.disabled) {
            return;
        }
        this.totalLen = this.sliderInner.nativeElement.clientWidth;
        this.start = this.percent * this.totalLen / 100;
        this.startX = event.targetTouches[0].clientX;
        this.onTouched();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    WeUISlider.prototype.onTouchMove = function (event) {
        if (this.disabled) {
            return;
        }
        var /** @type {?} */ dist = event.targetTouches[0].clientX - this.startX;
        var /** @type {?} */ end = this.start + dist;
        end = end < 0 ? 0 : end > this.totalLen ? this.totalLen : end;
        this.percent = end * 100 / this.totalLen;
        this.value = Math.round(this.min + (this.max - this.min) * this.percent / 100);
        event.preventDefault();
        this.onChange(this.value);
    };
    /**
     * Write a new value to the element. (From ControlValueAccessor interface)
     * @param {?} value
     * @return {?}
     */
    WeUISlider.prototype.writeValue = function (value) {
        if (value !== undefined && value !== null) {
            value = Math.min(Math.max(value, this.min), this.max);
            this.value = value;
            this.percent = (value + this.min) * 100 / this.max;
        }
    };
    /**
     * 设置禁用状态 (From ControlValueAccessor interface)
     * @param {?} isDisabled
     * @return {?}
     */
    WeUISlider.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    return WeUISlider;
}(DefaultValueAccessor));
WeUISlider.decorators = [
    { type: Component, args: [{
                selector: 'weui-slider',
                template: "\n        <div class=\"weui-slider-box weui-slider_{{color}}\">\n            <ng-content select=\"[weui-start]\"></ng-content>\n            <div class=\"weui-slider\">\n                <div #sliderInner class=\"weui-slider__inner\">\n                    <div [style.width]=\"percent + '%'\" class=\"weui-slider__track\"></div>\n                    <div [style.left]=\"percent + '%'\" class=\"weui-slider__handler\"\n                        (touchstart)=\"onTouchStart($event)\" (touchmove)=\"onTouchMove($event)\"></div>\n                </div>\n            </div>\n            <ng-content select=\"[weui-last]\"></ng-content>\n            <div class=\"weui-slider-box__value\" *ngIf=\"showValue\">{{value}}</div>\n        </div>\n    ",
                providers: [WEUI_FORM_CONTROL_VALUE_ACCESSOR$2]
            },] },
];
/**
 * @nocollapse
 */
WeUISlider.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUISlider.propDecorators = {
    'color': [{ type: Input },],
    'showValue': [{ type: Input },],
    'disabled': [{ type: Input },],
    'min': [{ type: Input },],
    'max': [{ type: Input },],
    'sliderInner': [{ type: ViewChild, args: ['sliderInner',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WEUI_FORM_CONTROL_VALUE_ACCESSOR$3 = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return WeUISwitch; }),
    multi: true
};
var WeUISwitch = (function (_super) {
    tslib_1.__extends(WeUISwitch, _super);
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} compositionMode
     */
    function WeUISwitch(renderer, elementRef, compositionMode) {
        var _this = _super.call(this, renderer, elementRef, compositionMode) || this;
        _this.renderer = renderer;
        _this.elementRef = elementRef;
        _this.compositionMode = compositionMode;
        /**
         * 是否已选中
         */
        _this.checked = false;
        /**
         * 基本样式
         */
        _this._cls_cell_switch = true;
        _this.value = 'on'; // default value
        return _this;
    }
    Object.defineProperty(WeUISwitch.prototype, "innerValue", {
        /**
         * The value of the input ngModel。 (view -> model)
         * @param {?} checked
         * @return {?}
         */
        set: function (checked) {
            if (this._value !== checked) {
                this._value = checked || false;
                // view -> model -> outside world (ie. NgModel on this control)
                this.onChange(this._value ? this.value : '');
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Write a new value to the element. (From ControlValueAccessor interface)
     * @param {?} value
     * @return {?}
     */
    WeUISwitch.prototype.writeValue = function (value) {
        this.checked = this.value === value;
        _super.prototype.writeValue.call(this, this.checked);
    };
    return WeUISwitch;
}(WeUIFormControl));
WeUISwitch.decorators = [
    { type: Component, args: [{
                selector: 'weui-switch',
                template: "\n        <div class=\"weui-cell__bd\">\n            {{label}}<ng-content select=\"label\"></ng-content>\n        </div>\n        <div class=\"weui-cell__ft\">\n            <label [for]=\"id\" class=\"weui-switch-cp\">\n                <input type=\"checkbox\" class=\"weui-switch-cp__input\"\n                    [attr.id]=\"id\" [attr.name]=\"name\" [value]=\"value\" [checked]=\"checked\" [(ngModel)]=\"innerValue\" />\n                <div class=\"weui-switch-cp__box\"></div>\n            </label>\n        </div>\n    ",
                providers: [WEUI_FORM_CONTROL_VALUE_ACCESSOR$3]
            },] },
];
/**
 * @nocollapse
 */
WeUISwitch.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COMPOSITION_BUFFER_MODE,] },] },
]; };
WeUISwitch.propDecorators = {
    '_cls_cell_switch': [{ type: HostBinding, args: ['class.weui-cell_switch',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUITabBarItem = (function () {
    /**
     * @param {?} tabbar
     */
    function WeUITabBarItem(tabbar) {
        this.tabbar = tabbar;
        /**
         * 设置基本样式
         */
        this.barItemCls = true;
        this._activated = false;
    }
    Object.defineProperty(WeUITabBarItem.prototype, "activatedCls", {
        /**
         * 设置激活样式
         * @return {?}
         */
        get: function () {
            return this._activated;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 点击触发激活
     *
     * @param {?} event 点击事件
     * @return {?}
     */
    WeUITabBarItem.prototype.onClick = function (event) {
        if (this.tabbar) {
            this.tabbar.activate(this);
        }
    };
    /**
     * 由父级控件传入
     *
     * @param {?} activated 激活状态
     * @return {?}
     */
    WeUITabBarItem.prototype.activated = function (activated) {
        this._activated = activated;
    };
    /**
     * @return {?}
     */
    WeUITabBarItem.prototype.isActivated = function () {
        return this._activated;
    };
    return WeUITabBarItem;
}());
WeUITabBarItem.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'weui-tabbar-item,[weui-tabbar-item]'
            },] },
];
/**
 * @nocollapse
 */
WeUITabBarItem.ctorParameters = function () { return [
    { type: WeUITabBar, decorators: [{ type: Optional }, { type: Host }, { type: Inject, args: [forwardRef(function () { return WeUITabBar; }),] },] },
]; };
WeUITabBarItem.propDecorators = {
    'value': [{ type: Input },],
    'barItemCls': [{ type: HostBinding, args: ['class.weui-tabbar__item',] },],
    'activatedCls': [{ type: HostBinding, args: ['class.weui-bar__item_on',] },],
    'onClick': [{ type: HostListener, args: ['click', ['$event'],] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUITabBar = (function () {
    function WeUITabBar() {
        /**
         * 初始激活的子对象
         */
        this.activeIndex = 0;
        /**
         * 设置基本样式
         */
        this.barCls = true;
        /**
         * 激活事件
         */
        this.activated = new EventEmitter();
    }
    /**
     * @return {?}
     */
    WeUITabBar.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.activateAt(_this.activeIndex);
        }, 1);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WeUITabBar.prototype.getAt = function (index) {
        return (this.items.find(function (item, i) { return i === index; }));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WeUITabBar.prototype.activate = function (item) {
        this._activated = item;
        this.items.forEach(function (child) {
            child.activated(child === item);
        });
        this.activated.emit(item);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WeUITabBar.prototype.activateAt = function (index) {
        var /** @type {?} */ item = this.getAt(index);
        if (item) {
            this.activate(item);
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WeUITabBar.prototype.isActivated = function (item) {
        return item.isActivated();
    };
    return WeUITabBar;
}());
WeUITabBar.decorators = [
    { type: Component, args: [{
                selector: 'weui-tabbar',
                template: "<ng-content></ng-content>"
            },] },
];
/**
 * @nocollapse
 */
WeUITabBar.ctorParameters = function () { return []; };
WeUITabBar.propDecorators = {
    'activeIndex': [{ type: Input },],
    'barCls': [{ type: HostBinding, args: ['class.weui-tabbar',] },],
    'items': [{ type: ContentChildren, args: [WeUITabBarItem,] },],
    'activated': [{ type: Output },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUITips = (function () {
    function WeUITips() {
        /**
         * 设置基本样式
         */
        this._cls_tips = true;
    }
    return WeUITips;
}());
WeUITips.decorators = [
    { type: Component, args: [{
                selector: 'weui-tips',
                template: "<ng-content></ng-content>"
            },] },
];
/**
 * @nocollapse
 */
WeUITips.ctorParameters = function () { return []; };
WeUITips.propDecorators = {
    '_cls_tips': [{ type: HostBinding, args: ['class.weui-cells__tips',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUITitle = (function () {
    function WeUITitle() {
        /**
         * 设置基本样式
         */
        this._cls_title = true;
    }
    return WeUITitle;
}());
WeUITitle.decorators = [
    { type: Component, args: [{
                selector: 'weui-title',
                template: "<ng-content></ng-content>"
            },] },
];
/**
 * @nocollapse
 */
WeUITitle.ctorParameters = function () { return []; };
WeUITitle.propDecorators = {
    '_cls_title': [{ type: HostBinding, args: ['class.weui-cells__title',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * 文件信息
 */
var WeUIFile = (function () {
    /**
     * @param {?} file
     * @param {?=} fileURL
     */
    function WeUIFile(file, fileURL) {
        /**
         * 此文件的上传进度
         */
        this.progress = 0;
        /**
         * 是否正在上传
         */
        this.isUploading = false;
        /**
         * 是否已经上传
         */
        this.isUploaded = false;
        this.file = file;
        this.fileURL = fileURL;
    }
    /**
     * 是否存在错误
     * @return {?}
     */
    WeUIFile.prototype.hasError = function () {
        return !!this.error;
    };
    /**
     * @return {?}
     */
    WeUIFile.prototype.hasMessage = function () {
        return this.hasError() || (this.progress > 0 && this.progress < 100);
    };
    /**
     * 重置状态<br>
     * 上传进度设置为0，错误消息为null，isUploading和isUploaded均设置为false。
     * @return {?}
     */
    WeUIFile.prototype.reset = function () {
        this.progress = 0;
        this.error = null;
        this.isUploading = false;
        this.isUploaded = false;
    };
    return WeUIFile;
}());
var WeUIUploader = (function () {
    /**
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} sanitizer
     */
    function WeUIUploader(renderer, elementRef, sanitizer) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.sanitizer = sanitizer;
        this.files = [];
    }
    /**
     * 获取列表中的文件
     * @return {?}
     */
    WeUIUploader.prototype.getFiles = function () {
        return this.files;
    };
    /**
     * 文件总数
     * @return {?}
     */
    WeUIUploader.prototype.getFileCount = function () {
        return this.files.length;
    };
    /**
     * 已上传文件数
     * @return {?}
     */
    WeUIUploader.prototype.getUploadedFileCount = function () {
        return this.files.filter(function (file) { return file.isUploaded; }).length;
    };
    /**
     * 选择图片文件
     * @param {?} event
     * @return {?}
     */
    WeUIUploader.prototype.onSelect = function (event) {
        var /** @type {?} */ picker = (event.target);
        var /** @type {?} */ files = picker && picker.files;
        if (!files) {
            return;
        }
        var /** @type {?} */ length = files.length;
        for (var /** @type {?} */ i = 0; i < length; i++) {
            var /** @type {?} */ file = files.item(i);
            var /** @type {?} */ url = 'url(' + window.URL.createObjectURL(file) + ')';
            var /** @type {?} */ safeUrl = this.sanitizer.bypassSecurityTrustStyle(url);
            this.files.push(new WeUIFile(file, safeUrl));
        }
    };
    /**
     * 预览图片
     * @param {?} file
     * @return {?}
     */
    WeUIUploader.prototype.preview = function (file) {
        this.image = file;
        if (this.gallery) {
            this.gallery.show();
        }
    };
    /**
     * 退出预览
     * @return {?}
     */
    WeUIUploader.prototype.fadeOut = function () {
        if (this.gallery) {
            this.gallery.hide();
        }
    };
    /**
     * 删除文件
     * @param {?} file
     * @return {?}
     */
    WeUIUploader.prototype.onDelete = function (file) {
        var /** @type {?} */ index = this.files.indexOf(file);
        if (index >= 0) {
            this.files.splice(index, 1);
        }
    };
    return WeUIUploader;
}());
WeUIUploader.decorators = [
    { type: Component, args: [{
                selector: 'weui-uploader',
                template: "\n        <div class=\"weui-uploader\">\n            <div class=\"weui-uploader__hd\">\n                <p class=\"weui-uploader__title\">\u56FE\u7247\u4E0A\u4F20</p>\n                <div class=\"weui-uploader__info\">{{getUploadedFileCount()}}/{{getFileCount()}}</div>\n            </div>\n            <div class=\"weui-uploader__bd\">\n                <ul class=\"weui-uploader__files\">\n                    <li class=\"weui-uploader__file\" *ngFor=\"let file of files\"\n                        [style.background-image]=\"file.fileURL\"\n                        [ngClass]=\"{'weui-uploader__file_status': file.hasMessage()}\"\n                        (click)=\"preview(file)\">\n                        <div class=\"weui-uploader__file-content\" *ngIf=\"file.hasMessage()\">\n                            <div class=\"weui-uploader__file-progress\"\n                                *ngIf=\"file.isUploading && file.progress && !file.hasError()\">{{file.progress}}%</div>\n                            <i class=\"weui-icon-warn\" *ngIf=\"file.hasError()\"></i>\n                        </div>\n                    </li>\n                </ul>\n                <div class=\"weui-uploader__input-box\">\n                    <input class=\"weui-uploader__input\" type=\"file\" accept=\"image/*\" multiple (change)=\"onSelect($event)\" />\n                </div>\n            </div>\n        </div>\n        <weui-gallery [image]=\"image\" [canDelete]=\"true\" (delete)=\"onDelete($event)\" (click)=\"fadeOut()\"></weui-gallery>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIUploader.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: DomSanitizer, },
]; };
WeUIUploader.propDecorators = {
    'gallery': [{ type: ViewChild, args: [forwardRef(function () { return WeUIGallery; }),] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIActionSheet = (function () {
    function WeUIActionSheet() {
        /**
         * \@i18n
         */
        this.defaults = {
            cancelText: '取消'
        };
        /**
         * ActionSheet菜单，列表，内容任意，其中<code>text</code>用于显示
         */
        this.menu = [];
        /**
         * ActionSheet弹出模式，取值：ios(Ios模式) - 从底部上弹，md(Android模式) - 弹出在窗口中间。默认为ios。
         */
        this.mode = 'ios';
        /**
         * \@i18n 取消
         */
        this.cancelText = this.defaults.cancelText;
        /**
         * 隐藏对象
         */
        this.close = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
        this._show = false;
    }
    Object.defineProperty(WeUIActionSheet.prototype, "androidCls", {
        /**
         * 安卓模式下的特殊样式
         * @return {?}
         */
        get: function () {
            return this.mode === 'md';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WeUIActionSheet.prototype, "hideCls", {
        /**
         * 用于控制控件的可视化
         * @return {?}
         */
        get: function () {
            return !this._show;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WeUIActionSheet.prototype, "visibility", {
        /**
         * 用于控制动画的触发(trigger)
         * @return {?}
         */
        get: function () {
            return this.shown ? 'show' : 'hide';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 显示菜单
     * @return {?}
     */
    WeUIActionSheet.prototype.show = function () {
        var _this = this;
        this._show = true;
        setTimeout(function () {
            _this.shown = true;
        }, 10);
        return new Promise(function (resolve, reject) {
            _this.resolve = resolve;
        });
    };
    /**
     * 隐藏菜单
     * @return {?}
     */
    WeUIActionSheet.prototype.hide = function () {
        var _this = this;
        this.shown = false;
        setTimeout(function () {
            _this._show = false;
            _this.close.emit();
        }, 200);
    };
    /**
     * 选择某个菜单，执行Promise.resolve()方法，并将选中的menu作为参数；最后，关闭ActionSheet
     *
     * @param {?} menu 被选择的菜单
     * @return {?}
     */
    WeUIActionSheet.prototype.onSelect = function (menu) {
        this.resolve(menu);
        this.hide();
    };
    return WeUIActionSheet;
}());
WeUIActionSheet.decorators = [
    { type: Component, args: [{
                selector: 'weui-actionsheet',
                template: "\n        <div class=\"weui-mask\" [@visibility]=\"visibility\" (click)=\"hide()\"></div>\n        <div class=\"weui-actionsheet\"\n            [ngClass]=\"{'weui-actionsheet_toggle': shown && mode == 'ios'}\">\n            <div class=\"weui-actionsheet__menu\">\n                <div class=\"weui-actionsheet__cell\" *ngFor=\"let m of menu\" (click)=\"onSelect(m)\">{{m.text}}</div>\n            </div>\n            <div class=\"weui-actionsheet__action\" *ngIf=\"mode == 'ios'\">\n                <div class=\"weui-actionsheet__cell\" (click)=\"hide()\">{{cancelText}}</div>\n            </div>\n        </div>\n    ",
                animations: [trigger('visibility', [
                        state('show', style({ opacity: 1 })),
                        state('hide', style({ opacity: 0 })),
                        transition('hide <=> show', [animate(200)])
                    ])]
            },] },
];
/**
 * @nocollapse
 */
WeUIActionSheet.ctorParameters = function () { return []; };
WeUIActionSheet.propDecorators = {
    'menu': [{ type: Input },],
    'mode': [{ type: Input },],
    'cancelText': [{ type: Input },],
    'close': [{ type: Output },],
    'androidCls': [{ type: HostBinding, args: ['class.weui-skin_android',] },],
    'hideCls': [{ type: HostBinding, args: ['class.weui-hide',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var WeUIDialog = (function () {
    function WeUIDialog() {
        /**
         * \@i18n
         */
        this.defaults = {
            btnNOText: '取消',
            btnOKText: '确定'
        };
        /**
         * ActionSheet弹出模式，取值：ios(Ios模式) - 从底部上弹，md(Android模式) - 弹出在窗口中间。默认为ios。
         */
        this.mode = 'ios';
        /**
         * \@i18n 取消
         */
        this.btnNOText = this.defaults.btnNOText;
        /**
         * \@i18n 确定
         */
        this.btnOKText = this.defaults.btnOKText;
        /**
         * 是否显示“取消”按钮
         */
        this.showNOButton = true;
        /**
         * 隐藏对象
         */
        this.close = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
    }
    Object.defineProperty(WeUIDialog.prototype, "visibility", {
        /**
         * 用于控制动画的触发(trigger)
         * @return {?}
         */
        get: function () {
            return this.shown ? 'show' : 'hide';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 显示对话框
     * @return {?}
     */
    WeUIDialog.prototype.show = function () {
        var _this = this;
        this.shown = true;
        return new Promise(function (resolve, reject) {
            _this.resolve = resolve;
            _this.reject = reject;
        });
    };
    /**
     * 隐藏对话框
     * @return {?}
     */
    WeUIDialog.prototype.hide = function () {
        this.shown = false;
        this.close.emit();
    };
    /**
     * 点击【取消】，执行Promise.reject()方法；然后，关闭对话框
     * @param {?} event
     * @return {?}
     */
    WeUIDialog.prototype.negativeClick = function (event) {
        this.reject();
        this.hide();
    };
    /**
     * 点击【确定】，执行Promise.resolve()方法；然后，关闭对话框
     * @param {?} event
     * @return {?}
     */
    WeUIDialog.prototype.positiveClick = function (event) {
        this.resolve();
        this.hide();
    };
    /**
     * @return {?}
     */
    WeUIDialog.prototype.getIcon = function () {
        if (this.icon) {
            return this.icon;
        }
        if (this.type) {
            return ({
                info: 'weui-icon-info',
                success: 'weui-icon-success',
                error: 'weui-icon-cancel',
                warning: 'weui-icon-warn',
                confirm: 'weui-icon-waiting',
                'default': 'weui-icon-info-circle',
            })[this.type];
        }
        return null;
    };
    /**
     * @return {?}
     */
    WeUIDialog.prototype.getIconClass = function () {
        var /** @type {?} */ icon = this.getIcon();
        return _a = {},
            _a["weui-dialog-icon"] = 1,
            _a["weui-dialog-icon-" + this.type] = this.type,
            _a["" + icon] = icon,
            _a["" + this.iconCls] = this.iconCls,
            _a;
        var _a;
    };
    /**
     * @return {?}
     */
    WeUIDialog.prototype.getWrapperClass = function () {
        return _a = {},
            _a["weui-dialog__wrapper"] = 1,
            _a["weui-dialog__wrapper_with_icon"] = this.getIcon(),
            _a;
        var _a;
    };
    return WeUIDialog;
}());
WeUIDialog.decorators = [
    { type: Component, args: [{
                selector: 'weui-dialog',
                template: "\n        <div class=\"weui-mask\"></div>\n        <div class=\"weui-dialog\" [ngClass]=\"{'weui-skin_android': mode == 'md'}\">\n            <div [ngClass]=\"getWrapperClass()\">\n                <i [ngClass]=\"getIconClass()\" *ngIf=\"getIcon()\"></i>\n                <div class=\"weui-dialog__wrapper_inner\">\n                    <div class=\"weui-dialog__hd\"><strong class=\"weui-dialog__title\">{{title}}</strong></div>\n                    <div class=\"weui-dialog__bd\">\n                        {{content}}\n                    </div>\n                </div>\n            </div>\n            <div class=\"weui-dialog__ft\">\n                <a href=\"javascript:;\" (click)=\"negativeClick($event)\" *ngIf=\"showNOButton\"\n                    class=\"weui-dialog__btn weui-dialog__btn_default\">{{btnNOText || defaults.btnNOText}}</a>\n                <a href=\"javascript:;\" (click)=\"positiveClick($event)\"\n                    class=\"weui-dialog__btn weui-dialog__btn_primary\">{{btnOKText || defaults.btnOKText}}</a>\n            </div>\n        </div>\n    ",
                animations: [trigger('visibility', [
                        state('show', style({ opacity: 1, display: 'block' })),
                        state('hide', style({ opacity: 0, display: 'none' })),
                        transition('hide <=> show', [animate(200)])
                    ])]
            },] },
];
/**
 * @nocollapse
 */
WeUIDialog.ctorParameters = function () { return []; };
WeUIDialog.propDecorators = {
    'mode': [{ type: Input },],
    'title': [{ type: Input },],
    'content': [{ type: Input },],
    'type': [{ type: Input },],
    'icon': [{ type: Input },],
    'iconCls': [{ type: Input },],
    'btnNOText': [{ type: Input },],
    'btnOKText': [{ type: Input },],
    'showNOButton': [{ type: Input },],
    'close': [{ type: Output },],
    'visibility': [{ type: HostBinding, args: ['@visibility',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * Toast - 弹出式提示
 */
var WeUIToast = (function () {
    /**
     * @param {?} renderer
     * @param {?} elementRef
     */
    function WeUIToast(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        /**
         * \@i18n
         */
        this.defaults = {
            loadingText: '数据加载中',
            successText: '操作成功'
        };
        /**
         * 状态：正在加载
         */
        this.loading = false;
        /**
         * 状态：操作成功
         */
        this.success = true;
        /**
         * 隐藏对象
         */
        this.close = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
    }
    Object.defineProperty(WeUIToast.prototype, "hideCls", {
        /**
         * 样式控制
         * @return {?}
         */
        get: function () {
            return !this.shown;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    WeUIToast.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ status = changes['status'];
        if (status) {
            if (status.currentValue === 'loading') {
                this.success = !(this.loading = true);
            }
            if (status.currentValue === 'success') {
                this.loading = !(this.success = true);
            }
        }
    };
    /**
     * 显示
     * @return {?}
     */
    WeUIToast.prototype.show = function () {
        var _this = this;
        this.shown = true;
        if (this.success) {
            setTimeout(function () { _this.hide(); }, 2000);
        }
    };
    /**
     * 隐藏
     * @return {?}
     */
    WeUIToast.prototype.hide = function () {
        this.shown = false;
        this.close.emit();
    };
    return WeUIToast;
}());
WeUIToast.decorators = [
    { type: Component, args: [{
                selector: 'weui-toast',
                template: "\n        <div class=\"weui-mask_transparent\"></div>\n        <div class=\"weui-toast\">\n            <i class=\"weui-icon-success-no-circle weui-icon_toast\" *ngIf=\"success\"></i>\n            <p class=\"weui-toast__content\" *ngIf=\"success\">{{ content || defaults.successText}}</p>\n            <i class=\"weui-loading weui-icon_toast\" *ngIf=\"loading\"></i>\n            <p class=\"weui-toast__content\" *ngIf=\"loading\">{{ content || defaults.loadingText}}</p>\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
WeUIToast.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
]; };
WeUIToast.propDecorators = {
    'content': [{ type: Input },],
    'status': [{ type: Input },],
    'loading': [{ type: Input },],
    'success': [{ type: Input },],
    'close': [{ type: Output },],
    'hideCls': [{ type: HostBinding, args: ['class.weui-hide',] },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * TopTips - 顶部提示
 */
var WeUITopTips = (function () {
    function WeUITopTips() {
        /**
         * 隐藏对象
         */
        this.close = new EventEmitter();
        /**
         * 已显示否
         */
        this.shown = false;
    }
    Object.defineProperty(WeUITopTips.prototype, "display", {
        /**
         * 样式控制
         * @return {?}
         */
        get: function () {
            return this.shown ? 'block' : 'none';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 显示
     * @return {?}
     */
    WeUITopTips.prototype.show = function () {
        var _this = this;
        this.shown = true;
        setTimeout(function () { _this.hide(); }, 2000);
    };
    /**
     * 隐藏
     * @return {?}
     */
    WeUITopTips.prototype.hide = function () {
        this.shown = false;
        this.close.emit();
    };
    return WeUITopTips;
}());
WeUITopTips.decorators = [
    { type: Component, args: [{
                selector: 'weui-toptips',
                template: '<div class="weui-toptips weui-toptips_warn" [ngStyle]="{display: display}">{{content}}</div>'
            },] },
];
/**
 * @nocollapse
 */
WeUITopTips.ctorParameters = function () { return []; };
WeUITopTips.propDecorators = {
    'content': [{ type: Input },],
    'close': [{ type: Output },],
};
/**
 * Next overlay unique ID.
 */
var nextUniqueId = 0;
/**
 * 浮层通用接口
 */
var Layer = (function () {
    /**
     * @param {?} _componentFactoryResolver
     * @param {?} _appRef
     * @param {?} _injector
     * @param {?} _ngZone
     */
    function Layer(_componentFactoryResolver, _appRef, _injector, _ngZone) {
        this._componentFactoryResolver = _componentFactoryResolver;
        this._appRef = _appRef;
        this._injector = _injector;
        this._ngZone = _ngZone;
    }
    /**
     * Creates the DOM element for an overlay and appends it to the document.
     * @return {?} Newly-created pane element
     */
    Layer.prototype._createDivElement = function () {
        var /** @type {?} */ div = document.createElement('div');
        div.id = "weui-overlay-" + nextUniqueId++;
        div.classList.add('weui-overlay-pane');
        document.body.appendChild(div);
        return div;
    };
    /**
     * @template T
     * @param {?} component
     * @return {?}
     */
    Layer.prototype._createCompInstance = function (component) {
        var /** @type {?} */ componentFactory = this._componentFactoryResolver.resolveComponentFactory(component);
        var /** @type {?} */ componentRef = componentFactory.create(this._injector);
        this._appRef.attachView(componentRef.hostView);
        // At this point the component has been instantiated, so we move it to the location in the DOM
        // where we want it to be rendered.
        var /** @type {?} */ _hostDomElement = this._createDivElement();
        _hostDomElement.appendChild(this._getComponentRootNode(componentRef));
        return componentRef;
    };
    /**
     * @return {?}
     */
    Layer.prototype._createDialog = function () {
        var /** @type {?} */ componentRef = this._createCompInstance(WeUIDialog);
        componentRef.instance.close.subscribe(function () {
            componentRef.destroy();
        });
        return componentRef.instance;
    };
    /**
     * 显示Success
     *
     * @param {?} message  消息
     * @param {?=} title    标题(可选)
     * @return {?}
     */
    Layer.prototype.success = function (message, title) {
        return this.showAlert(message, title, 'success');
    };
    /**
     * 显示Info
     *
     * @param {?} message  消息
     * @param {?=} title    标题(可选)
     * @return {?}
     */
    Layer.prototype.info = function (message, title) {
        return this.showAlert(message, title, 'info');
    };
    /**
     * 显示Error
     *
     * @param {?} message  消息
     * @param {?=} title    标题(可选)
     * @return {?}
     */
    Layer.prototype.error = function (message, title) {
        return this.showAlert(message, title, 'error');
    };
    /**
     * 显示Warning
     *
     * @param {?} message  消息
     * @param {?=} title    标题(可选)
     * @return {?}
     */
    Layer.prototype.warning = function (message, title) {
        return this.showAlert(message, title, 'warning');
    };
    /**
     * 显示Alert
     *
     * @param {?} message  消息
     * @param {?=} title    标题(可选)
     * @param {?=} type     告警等级
     * @return {?}
     */
    Layer.prototype.showAlert = function (message, title, type) {
        var /** @type {?} */ dialog = this._createDialog();
        dialog.showNOButton = false;
        if (typeof message === 'object') {
            for (var /** @type {?} */ key in message) {
                if (message.hasOwnProperty(key) && typeof message[key] !== 'undefined') {
                    dialog[key] = message[key];
                }
            }
        }
        else {
            dialog.content = message;
            dialog.title = title || '';
            dialog.type = type || null;
        }
        return dialog.show();
    };
    /**
     * 显示Confirm，默认带两个按钮（取消、确定）
     *
     * @param {?} message  消息，或者自定义对话框属性对象
     * @param {?=} title    标题
     * @return {?}
     */
    Layer.prototype.showConfirm = function (message, title) {
        var /** @type {?} */ dialog = this._createDialog();
        if (typeof message === 'object') {
            for (var /** @type {?} */ key in message) {
                if (message.hasOwnProperty(key) && typeof message[key] !== 'undefined') {
                    dialog[key] = message[key];
                }
            }
        }
        else {
            dialog.content = message;
            dialog.title = title || '';
        }
        return dialog.show();
    };
    /**
     * @return {?}
     */
    Layer.prototype._createToast = function () {
        var /** @type {?} */ componentRef = this._createCompInstance(WeUIToast);
        componentRef.instance.close.subscribe(function () {
            componentRef.destroy();
        });
        return componentRef.instance;
    };
    /**
     * 显示Loading Toast
     *
     * @param {?=} message  消息，默认为“数据加载中”
     * @return {?}
     */
    Layer.prototype.showLoading = function (message) {
        var /** @type {?} */ toast = this._createToast();
        if (message) {
            toast.content = message;
        }
        toast.success = !(toast.loading = true);
        toast.show();
        return toast;
    };
    /**
     * 显示成功Success Toast
     *
     * @param {?=} message  消息，默认为“操作成功”
     * @return {?}
     */
    Layer.prototype.showSuccess = function (message) {
        var /** @type {?} */ toast = this._createToast();
        if (message) {
            toast.content = message;
        }
        toast.loading = !(toast.success = true);
        toast.show();
        return toast;
    };
    /**
     * 在顶部显示错误消息
     *
     * @param {?} message  错误消息
     * @return {?}
     */
    Layer.prototype.showError = function (message) {
        var /** @type {?} */ componentRef = this._createCompInstance(WeUITopTips);
        componentRef.instance.close.subscribe(function () {
            componentRef.destroy();
        });
        var /** @type {?} */ tip = componentRef.instance;
        if (message) {
            tip.content = message;
        }
        tip.show();
    };
    /**
     * @return {?}
     */
    Layer.prototype._createActionSheet = function () {
        var /** @type {?} */ componentRef = this._createCompInstance(WeUIActionSheet);
        componentRef.instance.close.subscribe(function () {
            componentRef.destroy();
        });
        return componentRef.instance;
    };
    /**
     * 显示ActionSheet菜单
     *
     * @param {?} menus
     * @param {?=} btnCancelText 取消按钮文本，默认为“取消”
     * @return {?}
     */
    Layer.prototype.showActionsheet = function (menus, btnCancelText) {
        var /** @type {?} */ actionsheet = this._createActionSheet();
        actionsheet.menu = menus;
        if (btnCancelText) {
            actionsheet.cancelText = btnCancelText;
        }
        return actionsheet.show();
    };
    /**
     * Gets the root HTMLElement for an instantiated component.
     * @param {?} componentRef
     * @return {?}
     */
    Layer.prototype._getComponentRootNode = function (componentRef) {
        return (((componentRef.hostView)).rootNodes[0]);
    };
    return Layer;
}());
Layer.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
Layer.ctorParameters = function () { return [
    { type: ComponentFactoryResolver, },
    { type: ApplicationRef, },
    { type: Injector, },
    { type: NgZone, },
]; };
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
var seed = 0;
var now = Date.now();
/**
 * @return {?}
 */
function getUuid() {
    return "_picker_" + now + "_" + seed++;
}
/**
 * Picker服务
 */
var WeUIPickerService = (function () {
    /**
     * @param {?} _componentFactoryResolver
     * @param {?} _appRef
     * @param {?} _injector
     * @param {?} _ngZone
     * @param {?} doc
     */
    function WeUIPickerService(_componentFactoryResolver, _appRef, _injector, _ngZone, doc) {
        this._componentFactoryResolver = _componentFactoryResolver;
        this._appRef = _appRef;
        this._injector = _injector;
        this._ngZone = _ngZone;
        this.doc = doc;
    }
    /**
     * 显示单列Picker
     *
     * @param {?} menu  菜单选项
     * @return {?}
     */
    WeUIPickerService.prototype.showPicker = function (menu) {
        var /** @type {?} */ picker = this.createPicker(WeUIPicker, { menu: menu });
        return picker.show();
    };
    /**
     * 显示多列Picker
     *
     * @param {?} menus  菜单选项（多列）
     * @return {?}
     */
    WeUIPickerService.prototype.showMultiPicker = function (menus) {
        var /** @type {?} */ picker = this.createPicker(WeUIPicker, { menus: menus });
        return picker.show();
    };
    /**
     * 显示日期Picker
     *
     * @param {?=} config  配置项
     * @return {?}
     */
    WeUIPickerService.prototype.showDatePicker = function (config) {
        var /** @type {?} */ picker = this.createPicker(WeUIDatePicker, config);
        return picker.show();
    };
    /**
     * @param {?} component
     * @param {?} config
     * @return {?}
     */
    WeUIPickerService.prototype.createPicker = function (component, config) {
        var /** @type {?} */ componentRef = this._createComponent(component);
        var /** @type {?} */ instance = componentRef.instance;
        instance.close.subscribe(function () {
            componentRef.destroy();
            componentRef['_container'].remove();
        });
        if (config) {
            Object.assign(instance, config);
        }
        this.afterCreate(instance);
        return instance;
    };
    /**
     * this method is called before thild component's ngAfterViewInit
     * @param {?} instance
     * @return {?}
     */
    WeUIPickerService.prototype.afterCreate = function (instance) {
    };
    /**
     * @template T
     * @param {?} component
     * @return {?}
     */
    WeUIPickerService.prototype._createComponent = function (component) {
        var /** @type {?} */ componentFactory = this._componentFactoryResolver.resolveComponentFactory(component);
        var /** @type {?} */ componentRef = componentFactory.create(this._injector);
        this._appRef.attachView(componentRef.hostView);
        // At this point the component has been instantiated, so we move it to the location in the DOM
        // where we want it to be rendered.
        var /** @type {?} */ _hostDomElement = this._getContainer();
        _hostDomElement.appendChild(this._getComponentRootNode(componentRef));
        componentRef['_container'] = _hostDomElement;
        return componentRef;
    };
    /**
     * Gets the root HTMLElement for an instantiated component.
     * @param {?} componentRef
     * @return {?}
     */
    WeUIPickerService.prototype._getComponentRootNode = function (componentRef) {
        return (((componentRef.hostView)).rootNodes[0]);
    };
    /**
     * @return {?}
     */
    WeUIPickerService.prototype._getContainer = function () {
        var /** @type {?} */ div = ɵgetDOM().createElement('div');
        div.id = getUuid();
        div.classList.add('weui-overlay');
        ɵgetDOM().appendChild(this.doc.body, div);
        return div;
    };
    return WeUIPickerService;
}());
WeUIPickerService.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
WeUIPickerService.ctorParameters = function () { return [
    { type: ComponentFactoryResolver, },
    { type: ApplicationRef, },
    { type: Injector, },
    { type: NgZone, },
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
]; };
/**
 * @license
 * Let'say COFFEE © 2017-2018
 * Copyright 厦门美乐味餐饮管理有限公司 All Rights Reserved.
 */
/**
 * @param {?} obj
 * @return {?}
 */
function isPresent(obj) {
    return obj !== undefined && obj !== null;
}
var FormValidation = (function () {
    /**
     * @param {?} _element
     * @param {?} _form
     */
    function FormValidation(_element, _form) {
        this._element = _element;
        this._form = _form;
        /**
         * 自定义错误消息
         */
        this.messages = {};
        /**
         * 验证结果，例如: [{username: '必须填写'}, {password: '输入非法'}]
         */
        this.formErrors = [];
        this.defaultValidationMessages = {
            required: '该项为必填项',
            email: '请输入有效的电子邮件',
            url: '请输入有效的网址',
            date: '请输入有效的日期',
            dateISO: '请输入有效的日期 (YYYY-MM-DD)',
            number: '请输入正确的数字',
            digits: '只可输入数字',
            alphanumeric: '只可输入字母、数字及下划线',
            maxlength: '最多 {0} 个字符',
            minlength: '最少 {0} 个字符',
            rangelength: '请输入长度为 {0} 至 {1} 之间的字符',
            range: '请输入 {0} 至 {1} 之间的数值',
            max: '请输入不大于 {0} 的数值',
            min: '请输入不小于 {0} 的数值'
        };
        if (_form && _form.ngSubmit) {
            _form.ngSubmit.subscribe(this.onSubmit.bind(this));
        }
    }
    /**
     * @return {?}
     */
    FormValidation.prototype.showByName = function () {
        return isPresent(this.name) && this.name.length > 0;
    };
    /**
     * Form验证后，自动执行消息提示匹配
     * @return {?}
     */
    FormValidation.prototype.onSubmit = function () {
        this.validate();
    };
    /**
     * 验证表单，结果存入formErrors
     *
     * @return {?} 验证结果: true-success, false-fail
     */
    FormValidation.prototype.validate = function () {
        var /** @type {?} */ formErrors = [];
        var /** @type {?} */ form = this._form;
        if (!form) {
            return false;
        }
        for (var /** @type {?} */ field in form.controls) {
            if (form.controls.hasOwnProperty(field)) {
                var /** @type {?} */ control = form.controls[field];
                var /** @type {?} */ errors = control && control.errors || {};
                if (control && control.invalid && (control.dirty || form.submitted)) {
                    var /** @type {?} */ messages = (this.messages && this.messages[field]) || {};
                    if (errors['required']) {
                        formErrors.push({
                            name: field,
                            message: messages['required'] || this.defaultValidationMessages['required']
                        });
                        continue; // 默认先验证必填消息
                    }
                    for (var /** @type {?} */ key in errors) {
                        if (errors.hasOwnProperty(key)) {
                            formErrors.push({
                                name: field,
                                message: messages[key] || this.defaultValidationMessages[key]
                            });
                            continue; // 若一个输入框存在多个验证器，即使存在多个错误消息，也仅取一条错误消息
                        }
                    }
                }
            }
        }
        this.formErrors = formErrors;
        return form.valid === true;
    };
    Object.defineProperty(FormValidation.prototype, "submitted", {
        /**
         * 是否已submitted
         * @return {?}
         */
        get: function () {
            return this._form && this._form.submitted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormValidation.prototype, "valid", {
        /**
         * 是否验证通过
         * @return {?}
         */
        get: function () {
            return this._form && this._form.valid === true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormValidation.prototype, "invalid", {
        /**
         * 是否验证不通过
         * @return {?}
         */
        get: function () {
            return this._form && this._form.invalid === true;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 是否存在错误
     * @param {?=} name
     * @return {?}
     */
    FormValidation.prototype.hasError = function (name) {
        if (name) {
            return this.getError(name) !== null;
        }
        return this.formErrors.length > 0;
    };
    /**
     * 当存在多个错误消息时，用此方法获取其中的第一个非空错误显示
     * @return {?}
     */
    FormValidation.prototype.firstError = function () {
        var /** @type {?} */ err = this.formErrors[0];
        return (err && err.message) || null;
    };
    /**
     * 获取控件的一个错误
     *
     * @param {?} name 控件name
     * @return {?} 控件错误消息，或者空(没有错误)
     */
    FormValidation.prototype.getError = function (name) {
        var /** @type {?} */ len = this.formErrors.length;
        for (var /** @type {?} */ i = 0; i < len; i++) {
            var /** @type {?} */ err = this.formErrors[i];
            if (err.hasOwnProperty(name)) {
                return err[name];
            }
        }
        return null;
    };
    return FormValidation;
}());
FormValidation.decorators = [
    { type: Component, args: [{
                selector: 'form-validation',
                template: "\n        <div class=\"error\" *ngIf=\"showByName() && hasError(name)\">\n            {{getError(name)}}\n        </div>\n        <div class=\"error\" *ngIf=\"!showByName() && hasError()\">\n            {{firstError()}}\n        </div>\n    "
            },] },
];
/**
 * @nocollapse
 */
FormValidation.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: NgForm, },
]; };
FormValidation.propDecorators = {
    'messages': [{ type: Input },],
    'name': [{ type: Input },],
};
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * Import Angular
 */
/** Import WeUI Components */
/**
 * Import WeUI Overlay Entry Components
 */
/**
 * \@name WeUIModule
 * \@description
 * WeUIModule is an NgModule that provides weui-style TypeScript/JavaScript components.
 * @author fbchen 2017-03-08
 */
var WeUIModule = (function () {
    function WeUIModule() {
    }
    return WeUIModule;
}());
WeUIModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [
                    WeUIFormControl,
                    WeUIButton,
                    WeUISubmit,
                    WeUIInput,
                    WeUIRadio,
                    WeUICheckbox,
                    WeUISwitch,
                    WeUISelect,
                    WeUISlider,
                    WeUIGallery,
                    WeUIUploader,
                    WeUIItems,
                    WeUIItem,
                    WeUILink,
                    WeUITitle,
                    WeUITips,
                    WeUIBadge,
                    WeUIFooter,
                    WeUIFooterText,
                    WeUIFooterLinks,
                    WeUIFooterLink,
                    WeUILoadmore,
                    WeUIProgressBar,
                    WeUIActionSheet,
                    WeUIDialog,
                    WeUIToast,
                    WeUITopTips,
                    WeUIPicker,
                    WeUIDatePicker,
                    WeUIPickerGroup,
                    WeUISearchBar,
                    WeUINavBar,
                    WeUINavBarItem,
                    WeUITabBar,
                    WeUITabBarItem,
                    FormValidation
                ],
                exports: [
                    WeUIFormControl,
                    WeUIButton,
                    WeUISubmit,
                    WeUIInput,
                    WeUIRadio,
                    WeUICheckbox,
                    WeUISwitch,
                    WeUISelect,
                    WeUISlider,
                    WeUIGallery,
                    WeUIUploader,
                    WeUIItems,
                    WeUIItem,
                    WeUILink,
                    WeUITitle,
                    WeUITips,
                    WeUIBadge,
                    WeUIFooter,
                    WeUIFooterText,
                    WeUIFooterLinks,
                    WeUIFooterLink,
                    WeUILoadmore,
                    WeUIProgressBar,
                    WeUIActionSheet,
                    WeUIDialog,
                    WeUIToast,
                    WeUITopTips,
                    WeUIPicker,
                    WeUIDatePicker,
                    WeUISearchBar,
                    WeUINavBar,
                    WeUINavBarItem,
                    WeUITabBar,
                    WeUITabBarItem,
                    FormValidation
                ],
                entryComponents: [
                    WeUIActionSheet,
                    WeUIDialog,
                    WeUIToast,
                    WeUITopTips,
                    WeUIPicker,
                    WeUIDatePicker
                ],
                providers: [
                    Layer,
                    WeUIPickerService
                ]
            },] },
];
/**
 * @nocollapse
 */
WeUIModule.ctorParameters = function () { return []; };
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * @module
 * @description
 * Entry point for all public APIs of the weui package.
 */
/**
 * @license
 * Copyright 厦门乾元盛世科技有限公司 All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file.
 */
/**
 * @module
 * @description
 * Entry point for all public APIs of the weui package.
 */
// This file only reexports content of the `src` folder. Keep it that way.
/**
 * Generated bundle index. Do not edit.
 */
export { WeUIFormControl, WeUIBadge, WeUIButton, WeUISubmit, WeUICheckbox, WeUIDatePicker, WeUIFooter, WeUIFooterText, WeUIFooterLinks, WeUIFooterLink, WeUIGallery, WeUIInput, WeUIItems, WeUIItem, WeUILink, WeUILoadmore, WeUINavBar, WeUINavBarItem, WeUIPicker, WeUIPickerGroup, WeUIProgressBar, WeUIRadio, WeUISearchBar, WeUISelect, WeUISlider, WeUISwitch, WeUITabBar, WeUITabBarItem, WeUITips, WeUITitle, WeUIUploader, WeUIFile, WeUIActionSheet, WeUIDialog, WeUIToast, WeUITopTips, Layer, WeUIPickerService, FormValidation, WeUIModule };
